// @visactor/vchart/cjs/series/base/tooltip-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.BaseSeriesTooltipHelper = void 0;

const vutils_1 = require("@visactor/vutils"), tooltip_helper_1 = require("../../model/tooltip-helper"), common_1 = require("../../component/tooltip/utils/common");

class BaseSeriesTooltipHelper extends tooltip_helper_1.BaseTooltipHelper {
    constructor(series) {
        super(), this._getSeriesCacheInfo = () => {
            var _a, _b, _c;
            const {series: series} = this, _seriesField = series.getSeriesField();
            return {
                seriesFields: (0, vutils_1.isValid)(_seriesField) ? (0, vutils_1.array)(_seriesField) : null !== (_a = series.getSeriesKeys()) && void 0 !== _a ? _a : [],
                dimensionFields: null !== (_b = series.getDimensionField()) && void 0 !== _b ? _b : [],
                measureFields: null !== (_c = series.getMeasureField()) && void 0 !== _c ? _c : [],
                type: series.type
            };
        }, this._getDimensionData = datum => {
            const {dimensionFields: dimensionFields} = this._seriesCacheInfo;
            return dimensionFields[0] && (null == datum ? void 0 : datum[dimensionFields[0]]);
        }, this._getMeasureData = datum => {
            const {measureFields: measureFields} = this._seriesCacheInfo;
            return measureFields[0] && (null == datum ? void 0 : datum[measureFields[0]]);
        }, this._getSeriesStyle = (datum, styleKey, defaultValue) => {
            var _a;
            for (const key of (0, vutils_1.array)(styleKey)) {
                const value = null === (_a = this.series.getSeriesStyle(datum)) || void 0 === _a ? void 0 : _a(key);
                if ((0, vutils_1.isValid)(value)) return value;
            }
            return defaultValue;
        }, this.contentKeyCallback = (datum, params) => {
            const {dimensionFields: dimensionFields, seriesFields: seriesFields} = this._seriesCacheInfo, subDimensionField = dimensionFields[dimensionFields.length - 1];
            return (0, vutils_1.isValid)(seriesFields[0]) && (null == datum ? void 0 : datum[seriesFields[0]]) ? null == datum ? void 0 : datum[seriesFields[0]] : (dimensionFields.length > 1 && (0 === seriesFields.length || this.series.getSeriesKeys().length), 
            null == datum ? void 0 : datum[subDimensionField]);
        }, this.contentValueCallback = (datum, params) => this._getMeasureData(datum), this.contentShapeTypeCallback = (datum, params) => {
            var _a;
            return null !== (_a = this._getSeriesStyle(datum, "shape", null)) && void 0 !== _a ? _a : this._getSeriesStyle(datum, "symbolType", this.series.getDefaultShapeType());
        }, this.contentShapeColorCallback = (datum, params) => this._getSeriesStyle(datum, [ "fill", "stroke" ]), 
        this.titleValueCallback = (datum, params) => this._getDimensionData(datum), this.series = series, 
        this.updateTooltipSpec();
    }
    updateTooltipSpec() {
        var _a, _b, _c, _d;
        const seriesTooltipSpec = null !== (_b = null === (_a = this.series.getSpec()) || void 0 === _a ? void 0 : _a.tooltip) && void 0 !== _b ? _b : {}, chartTooltipSpec = null !== (_d = null === (_c = this.series.getChart().getSpec()) || void 0 === _c ? void 0 : _c.tooltip) && void 0 !== _d ? _d : {}, spec = Object.assign(Object.assign({}, chartTooltipSpec), seriesTooltipSpec);
        [ "mark", "dimension" ].forEach((activeType => {
            const pattern = spec[activeType];
            (0, vutils_1.isValid)(pattern) && (spec[activeType] = Object.assign(Object.assign({}, pattern), {
                title: (0, vutils_1.isValid)(pattern.title) ? (0, vutils_1.isFunction)(pattern.title) ? pattern.title : Object.assign(Object.assign({}, pattern.title), {
                    seriesId: this.series.id
                }) : void 0,
                content: (0, vutils_1.isValid)(pattern.content) ? (0, vutils_1.isFunction)(pattern.content) ? pattern.content : (0, 
                vutils_1.array)(pattern.content).map((line => (0, vutils_1.isFunction)(line) ? line : Object.assign(Object.assign({}, line), {
                    seriesId: this.series.id
                }))) : void 0
            }));
        })), this.spec = spec, this.activeType = (0, common_1.getTooltipActualActiveType)(spec), 
        this._seriesCacheInfo = this._getSeriesCacheInfo();
    }
    getDefaultTooltipPattern(activeType, dimensionInfo) {
        if ("mark" === activeType) return {
            visible: !0,
            activeType: activeType,
            title: {
                key: void 0,
                value: this.titleValueCallback,
                hasShape: !1
            },
            content: [ {
                seriesId: this.series.id,
                key: this.contentKeyCallback,
                value: this.contentValueCallback,
                hasShape: !0,
                shapeType: this.contentShapeTypeCallback,
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                shapeHollow: !1
            } ]
        };
        if ("dimension" === activeType && dimensionInfo) {
            const title = {
                key: void 0,
                value: this._getDimensionData,
                hasShape: !1
            }, content = [];
            return dimensionInfo.forEach((({data: data}) => data.forEach((({series: series}) => {
                content.push({
                    seriesId: series.id,
                    key: this.contentKeyCallback,
                    value: this.contentValueCallback,
                    hasShape: !0,
                    shapeType: this.contentShapeTypeCallback,
                    shapeColor: this.contentShapeColorCallback,
                    shapeStroke: this.contentShapeColorCallback,
                    shapeHollow: !1
                });
            })))), {
                visible: !0,
                activeType: activeType,
                title: title,
                content: content
            };
        }
        return null;
    }
}

exports.BaseSeriesTooltipHelper = BaseSeriesTooltipHelper;
