// @visactor/vchart/cjs/theme/builtin/common/series/funnel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.funnel = void 0;

const constants_1 = require("../constants");

exports.funnel = {
    transform: {
        style: {
            fill: {
                type: "palette",
                key: "axisGridColor"
            }
        }
    },
    label: {
        style: {
            fill: "white",
            textBaseline: "middle",
            lineWidth: 2
        }
    },
    outerLabel: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l4LineHeight,
            fill: {
                type: "palette",
                key: "secondaryFontColor"
            }
        },
        line: {
            style: {
                stroke: {
                    type: "palette",
                    key: "axisDomainColor"
                }
            }
        }
    },
    transformLabel: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l4LineHeight,
            fill: {
                type: "palette",
                key: "secondaryFontColor"
            },
            textBaseline: "middle"
        }
    }
};
