// @visactor/vchart/cjs/theme/builtin/common/series/funnel3d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.funnel3d = void 0;

const constants_1 = require("../constants");

exports.funnel3d = {
    transform3d: {
        style: {
            fill: "#f5f5f5"
        }
    },
    label: {
        style: {
            fill: "white",
            textBaseline: "middle",
            lineWidth: 2
        }
    },
    outerLabel: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l4LineHeight,
            fill: "#707070"
        },
        line: {
            style: {
                stroke: {
                    type: "palette",
                    key: "axisDomainColor"
                }
            }
        }
    },
    transformLabel: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l4LineHeight,
            fill: "#707070",
            textBaseline: "middle"
        }
    }
};
