// @visactor/vchart/cjs/theme/builtin/common/series/sunburst.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.sunburst = void 0;

const constant_1 = require("../../../../constant"), constants_1 = require("../constants");

exports.sunburst = {
    innerRadius: 0,
    outerRadius: 1,
    startAngle: constant_1.POLAR_START_ANGLE,
    endAngle: constant_1.POLAR_END_ANGLE,
    gap: 0,
    labelLayout: {
        align: "center",
        offset: 0,
        rotate: "radial"
    },
    sunburst: {
        style: {
            stroke: {
                type: "palette",
                key: "backgroundColor"
            },
            fillOpacity: 1,
            cursor: "pointer"
        }
    },
    label: {
        visible: !0,
        style: {
            cursor: "pointer",
            fill: {
                type: "palette",
                key: "primaryFontColor"
            },
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight
        }
    }
};
