// @visactor/vchart/cjs/theme/builtin/common/series/treemap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.treemap = void 0;

const constants_1 = require("../constants");

exports.treemap = {
    gapWidth: 1,
    nodePadding: [ 5 ],
    nonLeaf: {
        visible: !1,
        style: {
            fillOpacity: .5
        }
    },
    label: {
        style: {
            fill: "white",
            textBaseline: "middle",
            textAlign: "center",
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight
        }
    },
    nonLeafLabel: {
        padding: 2 * constants_1.THEME_CONSTANTS.l5FontSize,
        style: {
            fill: "black",
            stroke: {
                type: "palette",
                key: "backgroundColor"
            },
            lineWidth: 2,
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight,
            textBaseline: "middle",
            textAlign: "center"
        }
    }
};
