// @visactor/vchart/cjs/theme/theme-manager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.ThemeManager = void 0;

const builtin_1 = require("./builtin"), instance_manager_1 = require("../core/instance-manager");

class ThemeManager {
    static registerTheme(name, theme) {
        (0, builtin_1.registerTheme)(name, theme);
    }
    static getTheme(name, transformed = !1) {
        return (0, builtin_1.getTheme)(name, transformed);
    }
    static removeTheme(name) {
        return (0, builtin_1.removeTheme)(name);
    }
    static themeExist(name) {
        return (0, builtin_1.themeExist)(name);
    }
    static getDefaultTheme() {
        return ThemeManager.themes.get(builtin_1.defaultThemeName);
    }
    static setCurrentTheme(name) {
        ThemeManager.themeExist(name) && (ThemeManager._currentThemeName = name, instance_manager_1.InstanceManager.forEach((instance => null == instance ? void 0 : instance.setCurrentTheme(name))));
    }
    static getCurrentTheme() {
        return ThemeManager.getTheme(ThemeManager._currentThemeName);
    }
    static getCurrentThemeName() {
        return ThemeManager._currentThemeName;
    }
}

exports.ThemeManager = ThemeManager, ThemeManager.themes = builtin_1.themes, ThemeManager._currentThemeName = builtin_1.defaultThemeName;
