// @visactor/vchart/cjs/theme/builtin/light/color-scheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.colorScheme = void 0;

const default_1 = require("../../color-scheme/builtin/default");

exports.colorScheme = {
    default: {
        dataScheme: default_1.dataScheme,
        palette: {
            backgroundColor: "#ffffff",
            borderColor: "#e3e5e8",
            shadowColor: "rgba(33,37,44,0.1)",
            hoverBackgroundColor: "#f1f2f5",
            sliderRailColor: "#f1f3f4",
            sliderHandleColor: "#ffffff",
            sliderTrackColor: "#0040ff",
            popupBackgroundColor: "#ffffff",
            primaryFontColor: "#21252c",
            secondaryFontColor: "#606773",
            tertiaryFontColor: "#89909d",
            axisLabelFontColor: "#89909d",
            disableFontColor: "#bcc1cb",
            axisMarkerFontColor: "#ffffff",
            axisGridColor: "#f1f2f5",
            axisDomainColor: "#d9dde4",
            dataZoomHandleStrokeColor: "#aeb5be",
            dataZoomChartColor: "#c9ced8",
            playerControllerColor: "#0040ff",
            scrollBarSliderColor: "rgba(0,0,0,0.3)",
            axisMarkerBackgroundColor: "#21252c",
            markLabelBackgroundColor: "#f1f2f5",
            markLineStrokeColor: "#606773",
            dangerColor: "#e33232",
            warningColor: "#ffc528",
            successColor: "#07a35a",
            infoColor: "#3073f2"
        }
    }
};
