// @visactor/vchart/cjs/mark/label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerLabelMark = exports.LabelMark = void 0;

const vgrammar_core_1 = require("@visactor/vgrammar-core"), factory_1 = require("./../core/factory"), type_1 = require("./interface/type"), text_1 = require("./text");

class LabelMark extends text_1.TextMark {
    constructor() {
        super(...arguments), this.skipEncode = !1;
    }
    getRule() {
        return this._rule;
    }
    setRule(rule) {
        this._rule = rule;
    }
    getTarget() {
        return this._target;
    }
    setTarget(target) {
        this._target = target, this._rule || this.setRule(target.type);
    }
    getComponent() {
        return this._component;
    }
    setComponent(component) {
        this._component = component;
    }
}

exports.LabelMark = LabelMark, LabelMark.type = type_1.MarkTypeEnum.text, LabelMark.constructorType = type_1.MarkTypeEnum.label;

const registerLabelMark = () => {
    factory_1.Factory.registerMark(LabelMark.constructorType, LabelMark), (0, vgrammar_core_1.registerTextGraphic)();
};

exports.registerLabelMark = registerLabelMark;
