// @visactor/vchart/cjs/mark/rect-3d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRect3dMark = exports.Rect3dMark = void 0;

const factory_1 = require("./../core/factory"), base_mark_1 = require("./base/base-mark"), type_1 = require("./interface/type"), vgrammar_core_1 = require("@visactor/vgrammar-core"), config_1 = require("../animation/config");

class Rect3dMark extends base_mark_1.BaseMark {
    constructor() {
        super(...arguments), this.type = Rect3dMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            width: void 0,
            height: void 0,
            length: 3
        });
    }
}

exports.Rect3dMark = Rect3dMark, Rect3dMark.type = type_1.MarkTypeEnum.rect3d;

const registerRect3dMark = () => {
    factory_1.Factory.registerMark(Rect3dMark.type, Rect3dMark), (0, vgrammar_core_1.registerRect3dGraphic)(), 
    (0, config_1.registerVGrammarRectAnimation)();
};

exports.registerRect3dMark = registerRect3dMark;
