// @visactor/vchart/cjs/mark/rect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRectMark = exports.RectMark = void 0;

const factory_1 = require("./../core/factory"), base_mark_1 = require("./base/base-mark"), type_1 = require("./interface/type"), vgrammar_core_1 = require("@visactor/vgrammar-core"), config_1 = require("../animation/config");

class RectMark extends base_mark_1.BaseMark {
    constructor() {
        super(...arguments), this.type = RectMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            width: void 0,
            height: void 0,
            lineWidth: 0
        });
    }
}

exports.RectMark = RectMark, RectMark.type = type_1.MarkTypeEnum.rect;

const registerRectMark = () => {
    factory_1.Factory.registerMark(RectMark.type, RectMark), (0, vgrammar_core_1.registerRectGraphic)(), 
    (0, config_1.registerVGrammarRectAnimation)();
};

exports.registerRectMark = registerRectMark;
