// @visactor/vchart/cjs/mark/ripple.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRippleMark = exports.RippleMark = void 0;

const factory_1 = require("./../core/factory"), base_mark_1 = require("./base/base-mark"), type_1 = require("./interface/type"), vgrammar_core_1 = require("@visactor/vgrammar-core");

class RippleMark extends base_mark_1.BaseMark {
    constructor() {
        super(...arguments), this.type = RippleMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            x: 0,
            y: 0,
            ripple: 0
        });
    }
    _initProduct(group) {
        const view = this.getVGrammarView(), id = this.getProductId();
        this._product = view.glyph("ripplePoint", null != group ? group : view.rootMark).id(id), 
        this._compiledProductId = id;
    }
}

exports.RippleMark = RippleMark, RippleMark.type = type_1.MarkTypeEnum.ripple;

const registerRippleMark = () => {
    factory_1.Factory.registerMark(RippleMark.type, RippleMark), (0, vgrammar_core_1.registerRippleGlyph)();
};

exports.registerRippleMark = registerRippleMark;
