// @visactor/vchart/cjs/mark/symbol.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerSymbolMark = exports.SymbolMark = exports.BaseSymbolMark = void 0;

const factory_1 = require("./../core/factory"), typings_1 = require("../typings"), base_mark_1 = require("./base/base-mark"), type_1 = require("./interface/type"), vgrammar_core_1 = require("@visactor/vgrammar-core");

class BaseSymbolMark extends base_mark_1.BaseMark {
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            size: 1,
            symbolType: typings_1.ShapeTypeEnum.circle,
            fill: void 0,
            lineWidth: 0
        });
    }
}

exports.BaseSymbolMark = BaseSymbolMark;

class SymbolMark extends BaseSymbolMark {
    constructor() {
        super(...arguments), this.type = SymbolMark.type;
    }
}

exports.SymbolMark = SymbolMark, SymbolMark.type = type_1.MarkTypeEnum.symbol;

const registerSymbolMark = () => {
    factory_1.Factory.registerMark(SymbolMark.type, SymbolMark), (0, vgrammar_core_1.registerSymbolGraphic)();
};

exports.registerSymbolMark = registerSymbolMark;
