// @visactor/vchart/cjs/theme/builtin/common/mark.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.markByName = exports.markByType = void 0;

const constants_1 = require("./constants");

exports.markByType = {
    text: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l4LineHeight,
            fontWeight: "normal",
            fillOpacity: 1
        }
    }
}, exports.markByName = {
    label: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            fontWeight: "normal",
            fillOpacity: 1
        }
    },
    area: {
        style: {
            fillOpacity: constants_1.THEME_CONSTANTS.areaOpacity
        }
    },
    line: {
        style: {
            lineWidth: 2,
            lineCap: "round",
            lineJoin: "round"
        }
    },
    point: {
        style: {
            size: 8,
            stroke: {
                type: "palette",
                key: "backgroundColor"
            },
            lineWidth: 1,
            fillOpacity: 1
        }
    },
    word: {
        style: {
            fontSize: null
        }
    },
    fillingWord: {
        style: {
            fontSize: null
        }
    },
    sunburst: {
        style: {
            lineWidth: 1,
            stroke: {
                type: "palette",
                key: "backgroundColor"
            }
        }
    },
    circlePacking: {
        style: {
            lineWidth: 1,
            stroke: {
                type: "palette",
                key: "backgroundColor"
            }
        }
    },
    funnel3d: {
        style: {
            stroke: !1
        }
    },
    barBackground: {
        visible: !1,
        style: {
            fill: {
                type: "palette",
                key: "primaryFontColor",
                a: .06
            }
        }
    }
};
