// @visactor/vchart/cjs/series/pie/3d/pie-3d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerPie3dSeries = exports.Pie3dSeries = void 0;

const constant_1 = require("../../../constant"), type_1 = require("../../../mark/interface/type"), type_2 = require("../../interface/type"), arc_3d_1 = require("../../../mark/arc-3d"), pie_1 = require("../pie"), constant_2 = require("../constant"), vutils_1 = require("@visactor/vutils"), factory_1 = require("../../../core/factory"), animation_1 = require("../animation/animation"), pie_3d_transformer_1 = require("./pie-3d-transformer");

class Pie3dSeries extends pie_1.BasePieSeries {
    constructor() {
        super(...arguments), this.type = type_2.SeriesTypeEnum.pie3d, this._pieMarkName = type_2.SeriesMarkNameEnum.pie3d, 
        this._pieMarkType = type_1.MarkTypeEnum.arc3d, this.transformerConstructor = pie_3d_transformer_1.Pie3dSeriesSpecTransformer;
    }
    setAttrFromSpec() {
        var _a, _b;
        super.setAttrFromSpec(), this._angle3d = null !== (_b = null === (_a = this._spec) || void 0 === _a ? void 0 : _a.angle3d) && void 0 !== _b ? _b : -Math.PI / 3;
    }
    initMarkStyle() {
        super.initMarkStyle();
        const pieMark = this._pieMark;
        pieMark && this.setMarkStyle(pieMark, {
            beta: () => this._angle3d
        }, "normal", constant_1.AttributeLevel.Series);
    }
    initLabelMarkStyle(textMark) {
        if (!textMark) return;
        super.initLabelMarkStyle(textMark);
        const params3d = {};
        textMark.setSupport3d(!0), params3d.beta = this._angle3d, params3d.anchor3d = datum => {
            var _a, _b, _c, _d;
            return [ null !== (_b = null === (_a = this._center) || void 0 === _a ? void 0 : _a.x) && void 0 !== _b ? _b : this._region.getLayoutRect().width / 2, null !== (_d = null === (_c = this._center) || void 0 === _c ? void 0 : _c.y) && void 0 !== _d ? _d : this._region.getLayoutRect().height / 2 ];
        }, params3d.angle = datum => {
            var _a;
            const angle = datum[constant_1.ARC_MIDDLE_ANGLE];
            return "inside" === (null === (_a = this._spec.label) || void 0 === _a ? void 0 : _a.position) ? (0, 
            vutils_1.radianToDegree)(angle) : 0;
        }, this.setMarkStyle(textMark, Object.assign(Object.assign({}, params3d), {
            z: 100
        }), void 0, constant_1.AttributeLevel.Mark);
    }
}

exports.Pie3dSeries = Pie3dSeries, Pie3dSeries.type = type_2.SeriesTypeEnum.pie3d, 
Pie3dSeries.mark = constant_2.pie3dSeriesMark, Pie3dSeries.transformerConstructor = pie_3d_transformer_1.Pie3dSeriesSpecTransformer;

const registerPie3dSeries = () => {
    (0, animation_1.registerPie3dAnimation)(), (0, arc_3d_1.registerArc3dMark)(), factory_1.Factory.registerSeries(Pie3dSeries.type, Pie3dSeries);
};

exports.registerPie3dSeries = registerPie3dSeries;
