// @visactor/vgrammar-core/cjs/component/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerScrollbar = exports.registerGrid = exports.registerTitle = exports.registerPlayer = exports.registerDataZoom = exports.registerLabel = exports.registerSlider = exports.registerLegend = exports.registerAxis = void 0;

var axis_1 = require("./axis");

Object.defineProperty(exports, "registerAxis", {
    enumerable: !0,
    get: function() {
        return axis_1.registerAxis;
    }
});

var legend_1 = require("./legend");

Object.defineProperty(exports, "registerLegend", {
    enumerable: !0,
    get: function() {
        return legend_1.registerLegend;
    }
});

var slider_1 = require("./slider");

Object.defineProperty(exports, "registerSlider", {
    enumerable: !0,
    get: function() {
        return slider_1.registerSlider;
    }
});

var label_1 = require("./label");

Object.defineProperty(exports, "registerLabel", {
    enumerable: !0,
    get: function() {
        return label_1.registerLabel;
    }
});

var datazoom_1 = require("./datazoom");

Object.defineProperty(exports, "registerDataZoom", {
    enumerable: !0,
    get: function() {
        return datazoom_1.registerDataZoom;
    }
});

var player_1 = require("./player");

Object.defineProperty(exports, "registerPlayer", {
    enumerable: !0,
    get: function() {
        return player_1.registerPlayer;
    }
});

var title_1 = require("./title");

Object.defineProperty(exports, "registerTitle", {
    enumerable: !0,
    get: function() {
        return title_1.registerTitle;
    }
});

var grid_1 = require("./grid");

Object.defineProperty(exports, "registerGrid", {
    enumerable: !0,
    get: function() {
        return grid_1.registerGrid;
    }
});

var scrollbar_1 = require("./scrollbar");

Object.defineProperty(exports, "registerScrollbar", {
    enumerable: !0,
    get: function() {
        return scrollbar_1.registerScrollbar;
    }
});
