// @visactor/vrender-core/es/graphic/builtin-symbol/star.js
import { tau } from "@visactor/vutils";

import { BaseSymbol } from "./base";

const kr = Math.sin(Math.PI / 10) / Math.sin(7 * Math.PI / 10), kx = Math.sin(tau / 10) * kr, ky = -Math.cos(tau / 10) * kr;

export function star(ctx, r, transX, transY) {
    const x = kx * r, y = ky * r;
    ctx.moveTo(transX, -r + transY), ctx.lineTo(x + transX, y + transY);
    for (let i = 1; i < 5; ++i) {
        const a = tau * i / 5, c = Math.cos(a), s = Math.sin(a);
        ctx.lineTo(s * r + transX, -c * r + transY), ctx.lineTo(c * x - s * y + transX, s * x + c * y + transY);
    }
    return ctx.closePath(), !0;
}

export class StarSymbol extends BaseSymbol {
    constructor() {
        super(...arguments), this.type = "star", this.pathStr = "M4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,-4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,4.51351666838205,0";
    }
    draw(ctx, size, transX, transY) {
        return star(ctx, size / 2, transX, transY);
    }
    drawOffset(ctx, size, transX, transY, offset) {
        return star(ctx, size / 2 + offset, transX, transY);
    }
}

export default new StarSymbol;
