// @visactor/vchart/cjs/chart/circle-packing/circle-packing-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.CirclePackingChartSpecTransformer = void 0;

const interface_1 = require("../../series/interface"), base_1 = require("../base");

class CirclePackingChartSpecTransformer extends base_1.BaseChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            categoryField: spec.categoryField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            layoutPadding: spec.layoutPadding,
            label: spec.label,
            circlePacking: spec.circlePacking,
            drill: spec.drill,
            drillField: spec.drillField
        }), seriesType = interface_1.SeriesTypeEnum.circlePacking;
        return series.type = seriesType, series[seriesType] = spec[seriesType], series;
    }
    transformSpec(spec) {
        super.transformSpec(spec);
        const defaultSeriesSpec = this._getDefaultSeriesSpec(spec);
        spec.series && 0 !== spec.series.length ? spec.series.forEach((s => {
            this._isValidSeries(s.type) && Object.keys(defaultSeriesSpec).forEach((k => {
                k in s || (s[k] = defaultSeriesSpec[k]);
            }));
        })) : spec.series = [ defaultSeriesSpec ];
    }
}

exports.CirclePackingChartSpecTransformer = CirclePackingChartSpecTransformer;
