// @visactor/vchart-semi-theme/cjs/light/color-scheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.colorScheme = void 0;

const vchart_1 = require("@visactor/vchart");

exports.colorScheme = {
    default: {
        dataScheme: vchart_1.lightTheme.colorScheme.default.dataScheme,
        palette: {
            backgroundColor: "rgba(255,255,255,1)",
            borderColor: "rgba(28,31,35,0.08)",
            shadowColor: "rgba(0,0,0,0.1)",
            hoverBackgroundColor: "rgba(46,50,56,0.05)",
            sliderRailColor: "rgba(46,50,56,0.05)",
            sliderHandleColor: "rgba(255,255,255,1)",
            sliderTrackColor: "rgba(0,100,250,1)",
            popupBackgroundColor: "rgba(255,255,255,1)",
            primaryFontColor: "rgba(28,31,35,1)",
            secondaryFontColor: "rgba(28,31,35,0.8)",
            tertiaryFontColor: "rgba(28,31,35,0.62)",
            axisLabelFontColor: "rgba(28,31,35,0.62)",
            disableFontColor: "rgba(28,31,35,0.35)",
            axisMarkerFontColor: "rgba(255,255,255,1)",
            axisGridColor: "rgba(28,31,35,0.08)",
            axisDomainColor: "rgba(28,31,35,0.15)",
            dataZoomHandleStrokeColor: "rgba(46,50,56,0.13)",
            dataZoomChartColor: "rgba(46,50,56,0.09)",
            playerControllerColor: "rgba(0,100,250,1)",
            axisMarkerBackgroundColor: "rgba(28,31,35,1)",
            markLabelBackgroundColor: "rgba(28,31,35,0.08)",
            markLineStrokeColor: "rgba(28,31,35,0.8)",
            dangerColor: "rgba(249,57,32,1)",
            warningColor: "rgba(252,136,0,1)",
            successColor: "rgba(59,179,70,1)",
            infoColor: "rgba(0,100,250,1)"
        }
    }
};
