// @visactor/vchart/cjs/component/label/base-label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.BaseLabelComponent = void 0;

const base_component_1 = require("../base/base-component"), type_1 = require("../interface/type"), constant_1 = require("../../constant"), vutils_1 = require("@visactor/vutils");

class BaseLabelComponent extends base_component_1.BaseComponent {
    constructor(spec, options) {
        super(spec, options), this.type = type_1.ComponentTypeEnum.label, this.name = type_1.ComponentTypeEnum.label, 
        this.layoutType = "none", this.layoutZIndex = constant_1.LayoutZIndex.Label, this._regions = options.getRegionsInIndex([ options.specIndex ]);
    }
    _interactiveConfig(labelSpec) {
        const {interactive: interactive} = labelSpec;
        if (!0 !== interactive) return {
            hover: !1,
            select: !1
        };
        const result = {
            hover: !1,
            select: !1,
            state: labelSpec.state
        }, {hover: hover, select: select} = this._option.getChart().getSpec();
        return !1 === hover && !1 === hover.enable || (result.hover = !0), !1 === select && !1 === select.enable || (result.select = !0), 
        result;
    }
    _compareSpec(spec, prevSpec) {
        const result = super._compareSpec(spec, prevSpec);
        return result.reRender = !0, (0, vutils_1.isEqual)(prevSpec, spec) || (result.reMake = !0), 
        result;
    }
    onRender(ctx) {}
    changeRegions(regions) {}
    _getNeedClearVRenderComponents() {
        return [];
    }
}

exports.BaseLabelComponent = BaseLabelComponent, BaseLabelComponent.type = type_1.ComponentTypeEnum.label;
