// @visactor/vchart/cjs/series/range-column/tooltip-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.RangeColumnSeriesTooltipHelper = void 0;

const tooltip_helper_1 = require("../base/tooltip-helper"), space_1 = require("../../typings/space");

class RangeColumnSeriesTooltipHelper extends tooltip_helper_1.BaseSeriesTooltipHelper {
    getDefaultTooltipPattern(activeType) {
        return {
            visible: !0,
            activeType: activeType,
            title: {
                key: void 0,
                value: this.titleValueCallback,
                hasShape: !1
            },
            content: [ {
                key: this.contentKeyCallback,
                value: datum => this.series.getSpec().direction === space_1.Direction.horizontal ? datum[this.series.getSpec().xField[0]] + "-" + datum[this.series.getSpec().xField[1]] : datum[this.series.getSpec().yField[0]] + "-" + datum[this.series.getSpec().yField[1]],
                hasShape: !0,
                shapeType: this.contentShapeTypeCallback,
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                shapeHollow: !1
            } ]
        };
    }
}

exports.RangeColumnSeriesTooltipHelper = RangeColumnSeriesTooltipHelper;
