// @visactor/vgrammar-core/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.vglobal = exports.ThemeManager = exports.SIGNAL_VIEW_BOX = exports.SIGNAL_WIDTH = exports.SIGNAL_VIEW_WIDTH = exports.SIGNAL_VIEW_HEIGHT = exports.SIGNAL_PADDING = exports.SIGNAL_HEIGHT = exports.SIGNAL_AUTOFIT = exports.Factory = exports.GrammarBase = exports.invokeFunctionType = exports.parseFunctionType = exports.registerCoordinate = exports.Coordinate = exports.registerScale = exports.Scale = exports.View = exports.version = void 0, 
exports.version = "0.10.11", __exportStar(require("./graph"), exports), __exportStar(require("./interactions"), exports);

var view_1 = require("./view");

Object.defineProperty(exports, "View", {
    enumerable: !0,
    get: function() {
        return view_1.View;
    }
});

var scale_1 = require("./view/scale");

Object.defineProperty(exports, "Scale", {
    enumerable: !0,
    get: function() {
        return scale_1.Scale;
    }
}), Object.defineProperty(exports, "registerScale", {
    enumerable: !0,
    get: function() {
        return scale_1.registerScale;
    }
});

var coordinate_1 = require("./view/coordinate");

Object.defineProperty(exports, "Coordinate", {
    enumerable: !0,
    get: function() {
        return coordinate_1.Coordinate;
    }
}), Object.defineProperty(exports, "registerCoordinate", {
    enumerable: !0,
    get: function() {
        return coordinate_1.registerCoordinate;
    }
});

var util_1 = require("./parse/util");

Object.defineProperty(exports, "parseFunctionType", {
    enumerable: !0,
    get: function() {
        return util_1.parseFunctionType;
    }
}), Object.defineProperty(exports, "invokeFunctionType", {
    enumerable: !0,
    get: function() {
        return util_1.invokeFunctionType;
    }
});

var grammar_base_1 = require("./view/grammar-base");

Object.defineProperty(exports, "GrammarBase", {
    enumerable: !0,
    get: function() {
        return grammar_base_1.GrammarBase;
    }
}), __exportStar(require("./util/text"), exports), __exportStar(require("./types"), exports);

var factory_1 = require("./core/factory");

Object.defineProperty(exports, "Factory", {
    enumerable: !0,
    get: function() {
        return factory_1.Factory;
    }
});

var constants_1 = require("./view/constants");

Object.defineProperty(exports, "SIGNAL_AUTOFIT", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_AUTOFIT;
    }
}), Object.defineProperty(exports, "SIGNAL_HEIGHT", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_HEIGHT;
    }
}), Object.defineProperty(exports, "SIGNAL_PADDING", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_PADDING;
    }
}), Object.defineProperty(exports, "SIGNAL_VIEW_HEIGHT", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_VIEW_HEIGHT;
    }
}), Object.defineProperty(exports, "SIGNAL_VIEW_WIDTH", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_VIEW_WIDTH;
    }
}), Object.defineProperty(exports, "SIGNAL_WIDTH", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_WIDTH;
    }
}), Object.defineProperty(exports, "SIGNAL_VIEW_BOX", {
    enumerable: !0,
    get: function() {
        return constants_1.SIGNAL_VIEW_BOX;
    }
});

var theme_manager_1 = require("./theme/theme-manager");

Object.defineProperty(exports, "ThemeManager", {
    enumerable: !0,
    get: function() {
        return theme_manager_1.ThemeManager;
    }
}), __exportStar(require("./glyph"), exports), __exportStar(require("./component"), exports), 
__exportStar(require("./transforms"), exports), __exportStar(require("./graph/animation/animation"), exports);

var vrender_core_1 = require("@visactor/vrender-core");

Object.defineProperty(exports, "vglobal", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.vglobal;
    }
}), __exportStar(require("./env"), exports), __exportStar(require("./interactions"), exports), 
__exportStar(require("./semantic-marks/cell"), exports), __exportStar(require("./semantic-marks/interval"), exports), 
__exportStar(require("./graph/mark/graphic"), exports);
