// @visactor/vchart/esm/chart/sunburst/sunburst-transformer.js
import { isValid, radianToDegree } from "@visactor/vutils";

import { BaseChartSpecTransformer } from "../base";

import { POLAR_START_ANGLE } from "../../constant";

import { SeriesTypeEnum } from "../../series";

export class SunburstChartSpecTransformer extends BaseChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const startAngle = isValid(spec.startAngle) ? spec.startAngle : POLAR_START_ANGLE, endAngle = isValid(spec.endAngle) ? spec.endAngle : startAngle + radianToDegree(2 * Math.PI), series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            categoryField: spec.categoryField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            centerX: spec.centerX,
            centerY: spec.centerY,
            offsetX: spec.offsetX,
            offsetY: spec.offsetY,
            startAngle: startAngle,
            endAngle: endAngle,
            innerRadius: spec.innerRadius,
            outerRadius: spec.outerRadius,
            gap: spec.gap,
            labelLayout: spec.labelLayout,
            label: spec.label,
            labelAutoVisible: spec.labelAutoVisible,
            drill: spec.drill,
            drillField: spec.drillField
        }), seriesType = SeriesTypeEnum.sunburst;
        return series.type = seriesType, series[seriesType] = spec[seriesType], series;
    }
    transformSpec(spec) {
        super.transformSpec(spec);
        const defaultSeriesSpec = this._getDefaultSeriesSpec(spec);
        spec.series && 0 !== spec.series.length ? spec.series.forEach((s => {
            this._isValidSeries(s.type) && Object.keys(defaultSeriesSpec).forEach((k => {
                k in s || (s[k] = defaultSeriesSpec[k]);
            }));
        })) : spec.series = [ defaultSeriesSpec ];
    }
}
