// @visactor/vrender-core/cjs/graphic/builtin-symbol/rect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.RectSymbol = exports.rectSize = exports.rectSizeArray = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("./base");

function rectSizeArray(ctx, size, x, y) {
    return ctx.rect(x - size[0] / 2, y - size[1] / 2, size[0], size[1]), !1;
}

function rectSize(ctx, size, x, y) {
    const w = size, h = size / 2;
    return ctx.rect(x - w / 2, y - h / 2, w, h), !1;
}

exports.rectSizeArray = rectSizeArray, exports.rectSize = rectSize;

class RectSymbol extends base_1.BaseSymbol {
    constructor() {
        super(...arguments), this.type = "rect", this.pathStr = "M -0.5,0.25 L 0.5,0.25 L 0.5,-0.25,L -0.5,-0.25 Z";
    }
    draw(ctx, size, x, y) {
        return (0, vutils_1.isNumber)(size) ? rectSize(ctx, size, x, y) : rectSizeArray(ctx, size, x, y);
    }
    drawOffset(ctx, size, x, y, offset) {
        return (0, vutils_1.isNumber)(size) ? rectSize(ctx, size + 2 * offset, x, y) : rectSizeArray(ctx, [ size[0] + 2 * offset, size[1] + 2 * offset ], x, y);
    }
}

exports.RectSymbol = RectSymbol, exports.default = new RectSymbol;
