// @visactor/vrender-core/cjs/graphic/builtin-symbol/star.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.StarSymbol = exports.star = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("./base"), kr = Math.sin(Math.PI / 10) / Math.sin(7 * Math.PI / 10), kx = Math.sin(vutils_1.tau / 10) * kr, ky = -Math.cos(vutils_1.tau / 10) * kr;

function star(ctx, r, transX, transY) {
    const x = kx * r, y = ky * r;
    ctx.moveTo(transX, -r + transY), ctx.lineTo(x + transX, y + transY);
    for (let i = 1; i < 5; ++i) {
        const a = vutils_1.tau * i / 5, c = Math.cos(a), s = Math.sin(a);
        ctx.lineTo(s * r + transX, -c * r + transY), ctx.lineTo(c * x - s * y + transX, s * x + c * y + transY);
    }
    return ctx.closePath(), !0;
}

exports.star = star;

class StarSymbol extends base_1.BaseSymbol {
    constructor() {
        super(...arguments), this.type = "star", this.pathStr = "M4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,-4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,4.51351666838205,0";
    }
    draw(ctx, size, transX, transY) {
        return star(ctx, size / 2, transX, transY);
    }
    drawOffset(ctx, size, transX, transY, offset) {
        return star(ctx, size / 2 + offset, transX, transY);
    }
}

exports.StarSymbol = StarSymbol, exports.default = new StarSymbol;
