// @visactor/vrender-core/cjs/graphic/builtin-symbol/triangle-up.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.TriangleUpSymbol = exports.trianglUpOffset = exports.trianglUp = void 0;

const base_1 = require("./base");

function trianglUp(ctx, r, x, y) {
    return ctx.moveTo(x + r, r + y), ctx.lineTo(x - r, r + y), ctx.lineTo(x, y - r), 
    ctx.closePath(), !0;
}

function trianglUpOffset(ctx, r, x, y, offset) {
    return ctx.moveTo(x + r + 2 * offset, r + y + offset), ctx.lineTo(x - r - 2 * offset, r + y + offset), 
    ctx.lineTo(x, y - r - 2 * offset), ctx.closePath(), !0;
}

exports.trianglUp = trianglUp, exports.trianglUpOffset = trianglUpOffset;

class TriangleUpSymbol extends base_1.BaseSymbol {
    constructor() {
        super(...arguments), this.type = "triangleUp", this.pathStr = "M0.5,0.5 L-0.5,0.5 L0,-0.5 Z";
    }
    draw(ctx, size, x, y) {
        return trianglUp(ctx, size / 2, x, y);
    }
    drawOffset(ctx, size, x, y, offset) {
        return trianglUpOffset(ctx, size / 2, x, y, offset);
    }
}

exports.TriangleUpSymbol = TriangleUpSymbol, exports.default = new TriangleUpSymbol;
