// @visactor/vrender-core/cjs/graphic/builtin-symbol/triangle-down.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.TriangleDownSymbol = exports.trianglDownOffset = exports.trianglDown = void 0;

const base_1 = require("./base");

function trianglDown(ctx, r, x, y) {
    return ctx.moveTo(x - r, y - r), ctx.lineTo(x + r, y - r), ctx.lineTo(x, y + r), 
    ctx.closePath(), !0;
}

function trianglDownOffset(ctx, r, x, y, offset) {
    return ctx.moveTo(x - r - 2 * offset, y - r - offset), ctx.lineTo(x + r + 2 * offset, y - r - offset), 
    ctx.lineTo(x, y + r + 2 * offset), ctx.closePath(), !0;
}

exports.trianglDown = trianglDown, exports.trianglDownOffset = trianglDownOffset;

class TriangleDownSymbol extends base_1.BaseSymbol {
    constructor() {
        super(...arguments), this.type = "triangleDown", this.pathStr = "M-0.5,-0.5 L0.5,-0.5 L0,0.5 Z";
    }
    draw(ctx, size, x, y) {
        return trianglDown(ctx, size / 2, x, y);
    }
    drawOffset(ctx, size, x, y, offset) {
        return trianglDownOffset(ctx, size / 2, x, y, offset);
    }
}

exports.TriangleDownSymbol = TriangleDownSymbol, exports.default = new TriangleDownSymbol;
