// @visactor/vrender-core/cjs/graphic/builtin-symbol/wye.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.WyeSymbol = exports.wye = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("./base"), c = -.5, s = (0, 
vutils_1.sqrt)(3) / 2, k = 1 / (0, vutils_1.sqrt)(12), a = 3 * (k / 2 + 1);

function wye(ctx, r, transX, transY) {
    const x0 = r / 2, y0 = r * k, x1 = x0, y1 = r * k + r, x2 = -x1, y2 = y1;
    return ctx.moveTo(x0 + transX, y0 + transY), ctx.lineTo(x1 + transX, y1 + transY), 
    ctx.lineTo(x2 + transX, y2 + transY), ctx.lineTo(c * x0 - s * y0 + transX, s * x0 + c * y0 + transY), 
    ctx.lineTo(c * x1 - s * y1 + transX, s * x1 + c * y1 + transY), ctx.lineTo(c * x2 - s * y2 + transX, s * x2 + c * y2 + transY), 
    ctx.lineTo(c * x0 + s * y0 + transX, c * y0 - s * x0 + transY), ctx.lineTo(c * x1 + s * y1 + transX, c * y1 - s * x1 + transY), 
    ctx.lineTo(c * x2 + s * y2 + transX, c * y2 - s * x2 + transY), ctx.closePath(), 
    !1;
}

exports.wye = wye;

class WyeSymbol extends base_1.BaseSymbol {
    constructor() {
        super(...arguments), this.type = "wye", this.pathStr = "M4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,-4.51351666838205,0A4.51351666838205,4.51351666838205,0,1,1,4.51351666838205,0";
    }
    draw(ctx, size, transX, transY) {
        return wye(ctx, size / 2, transX, transY);
    }
    drawOffset(ctx, size, transX, transY, offset) {
        return wye(ctx, size / 2 + offset, transX, transY);
    }
}

exports.WyeSymbol = WyeSymbol, exports.default = new WyeSymbol;
