// @visactor/vchart/esm/mark/text.js
import { Factory } from "./../core/factory";

import { BaseMark } from "./base/base-mark";

import { MarkTypeEnum } from "./interface";

import { GrammarMarkType, registerRichTextGraphic, registerTextGraphic } from "@visactor/vgrammar-core";

export class TextMark extends BaseMark {
    constructor() {
        super(...arguments), this.type = TextMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            angle: 0,
            textAlign: "center",
            lineWidth: 0,
            textConfig: []
        });
    }
    _initProduct(group) {
        const view = this.getVGrammarView(), id = this.getProductId();
        "rich" === this.getStyle("textType") ? this._product = view.mark(GrammarMarkType.richtext, null != group ? group : view.rootMark).id(id) : this._product = view.mark(GrammarMarkType.text, null != group ? group : view.rootMark).id(id), 
        this._compiledProductId = id;
    }
}

TextMark.type = MarkTypeEnum.text;

export const registerTextMark = () => {
    Factory.registerMark(TextMark.type, TextMark), registerTextGraphic(), registerRichTextGraphic();
};
