// @visactor/vchart/esm/chart/circle-packing/circle-packing-transformer.js
import { SeriesTypeEnum } from "../../series/interface";

import { BaseChartSpecTransformer } from "../base";

export class CirclePackingChartSpecTransformer extends BaseChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            categoryField: spec.categoryField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            layoutPadding: spec.layoutPadding,
            label: spec.label,
            circlePacking: spec.circlePacking,
            drill: spec.drill,
            drillField: spec.drillField
        }), seriesType = SeriesTypeEnum.circlePacking;
        return series.type = seriesType, series[seriesType] = spec[seriesType], series;
    }
    transformSpec(spec) {
        super.transformSpec(spec);
        const defaultSeriesSpec = this._getDefaultSeriesSpec(spec);
        spec.series && 0 !== spec.series.length ? spec.series.forEach((s => {
            this._isValidSeries(s.type) && Object.keys(defaultSeriesSpec).forEach((k => {
                k in s || (s[k] = defaultSeriesSpec[k]);
            }));
        })) : spec.series = [ defaultSeriesSpec ];
    }
}
