// @visactor/vchart/esm/series/sankey/animation.js
import { Direction } from "../../typings/space";

import { Factory } from "../../core/factory";

import { FadeInOutAnimation } from "../../animation/config";

export const sankeyGrowIn = (params, isOverall = !0) => ({
    type: params.direction === Direction.horizontal ? "growWidthIn" : "growHeightIn",
    options: {
        overall: isOverall ? params.growFrom() : isOverall,
        orient: Direction.horizontal ? "positive" : "negative"
    }
});

export const sankeyGrowOut = (params, isOverall = !0) => ({
    type: params.direction === Direction.horizontal ? "growWidthOut" : "growHeightOut",
    options: {
        overall: isOverall ? params.growFrom() : isOverall,
        orient: Direction.horizontal ? "positive" : "negative"
    }
});

export const sankeyNodePresetAnimation = (params, preset) => "fadeIn" === preset ? {
    type: "fadeIn"
} : sankeyGrowIn(params);

export const sankeyLinkPresetAnimation = preset => "fadeIn" === preset ? {
    type: "fadeIn"
} : {
    type: "linkPathGrowIn"
};

export const registerSankeyAnimation = () => {
    Factory.registerAnimation("sankeyNode", ((params, preset) => Object.assign({
        appear: sankeyNodePresetAnimation(params, preset)
    }, FadeInOutAnimation))), Factory.registerAnimation("sankeyLinkPath", ((params, preset) => ({
        appear: sankeyLinkPresetAnimation(preset),
        enter: {
            type: "linkPathGrowIn"
        },
        exit: {
            type: "linkPathGrowOut"
        },
        disappear: {
            type: "linkPathGrowOut"
        }
    })));
};
