// @visactor/vchart/cjs/chart/progress/circular/circular-progress-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.CircularProgressChartSpecTransformer = void 0;

const polar_1 = require("../../polar");

class CircularProgressChartSpecTransformer extends polar_1.ProgressLikeChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        var _a, _b;
        const series = super._getDefaultSeriesSpec(spec);
        return Object.assign(Object.assign({}, series), {
            cornerRadius: null !== (_a = spec.cornerRadius) && void 0 !== _a ? _a : 0,
            roundCap: null !== (_b = spec.roundCap) && void 0 !== _b && _b,
            progress: spec.progress,
            track: spec.track,
            tickMask: spec.tickMask
        });
    }
    transformSpec(spec) {
        super.transformSpec(spec), this._transformProgressAxisSpec(spec, {
            orient: "angle",
            visible: !1
        }, {
            orient: "radius",
            visible: !1
        });
    }
}

exports.CircularProgressChartSpecTransformer = CircularProgressChartSpecTransformer;
