// @visactor/vchart/cjs/series/area/animation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerAreaAnimation = exports.areaPresetAnimation = void 0;

const config_1 = require("../../animation/config"), factory_1 = require("../../core/factory"), space_1 = require("../../typings/space"), Appear_ClipIn = {
    type: "clipIn"
}, Appear_Grow = params => ({
    type: params.direction === space_1.Direction.horizontal ? "growPointsXIn" : "growPointsYIn",
    options: {
        orient: params.direction === space_1.Direction.horizontal ? "positive" : "negative"
    }
}), Appear_FadeIn = {
    type: "fadeIn"
};

function areaPresetAnimation(params, preset) {
    if (!1 === preset) return {};
    switch (preset) {
      case "grow":
        return Appear_Grow(params);

      case "fadeIn":
        return Appear_FadeIn;

      default:
        return Appear_ClipIn;
    }
}

exports.areaPresetAnimation = areaPresetAnimation;

const registerAreaAnimation = () => {
    factory_1.Factory.registerAnimation("area", ((params, preset) => ({
        appear: areaPresetAnimation(params, preset),
        enter: {
            type: "fadeIn"
        },
        exit: {
            type: "fadeOut"
        },
        disappear: {
            type: "clipOut"
        }
    }))), (0, config_1.registerLineAnimation)(), (0, config_1.registerScaleInOutAnimation)();
};

exports.registerAreaAnimation = registerAreaAnimation;
