// @visactor/vchart/cjs/theme/builtin/common/component/crosshair.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.crosshair = void 0;

const constants_1 = require("../constants"), getLabelSpec = () => ({
    visible: !1,
    style: {
        fontWeight: "normal",
        fill: {
            type: "palette",
            key: "axisMarkerFontColor"
        },
        fontSize: constants_1.THEME_CONSTANTS.l5FontSize
    },
    labelBackground: {
        padding: {
            bottom: 0,
            top: 0,
            left: 2,
            right: 2
        },
        style: {
            fill: {
                type: "palette",
                key: "axisMarkerBackgroundColor"
            },
            cornerRadius: 1
        }
    }
}), getBandField = () => ({
    visible: !1,
    line: {
        type: "rect",
        visible: !0,
        style: {
            fill: {
                type: "palette",
                key: "axisGridColor"
            },
            opacity: .7,
            lineDash: []
        }
    },
    label: getLabelSpec()
}), getLinearField = () => ({
    visible: !1,
    line: {
        type: "line",
        visible: !0,
        style: {
            stroke: {
                type: "palette",
                key: "markLineStrokeColor"
            },
            fill: {
                type: "palette",
                key: "axisGridColor"
            },
            opacity: .7,
            lineDash: [ 2, 3 ]
        }
    },
    label: getLabelSpec()
});

exports.crosshair = {
    trigger: "hover",
    bandField: {
        visible: !1,
        line: {
            type: "rect",
            visible: !0,
            style: {
                fill: {
                    type: "palette",
                    key: "axisGridColor"
                },
                opacity: .7,
                lineDash: []
            }
        },
        label: getLabelSpec()
    },
    linearField: {
        visible: !1,
        line: {
            type: "line",
            visible: !0,
            style: {
                stroke: {
                    type: "palette",
                    key: "markLineStrokeColor"
                },
                fill: {
                    type: "palette",
                    key: "axisGridColor"
                },
                opacity: .7,
                lineDash: [ 2, 3 ]
            }
        },
        label: getLabelSpec()
    }
};
