// @visactor/vchart/cjs/theme/builtin/common/component/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.component = void 0;

const band_axis_1 = require("./axis/band-axis"), cartesian_axis_1 = require("./axis/cartesian-axis"), common_axis_1 = require("./axis/common-axis"), linear_axis_1 = require("./axis/linear-axis"), polar_axis_1 = require("./axis/polar-axis"), brush_1 = require("./brush"), crosshair_1 = require("./crosshair"), data_zoom_1 = require("./data-zoom"), indicator_1 = require("./indicator"), color_legend_1 = require("./legend/color-legend"), discrete_legend_1 = require("./legend/discrete-legend"), size_legend_1 = require("./legend/size-legend"), map_label_1 = require("./map-label"), mark_area_1 = require("./mark-area"), mark_line_1 = require("./mark-line"), mark_point_1 = require("./mark-point"), player_1 = require("./player"), title_1 = require("./title"), tooltip_1 = require("./tooltip"), poptip_1 = require("./poptip"), total_label_1 = require("./total-label"), scroll_bar_1 = require("./scroll-bar");

exports.component = {
    discreteLegend: discrete_legend_1.discreteLegend,
    colorLegend: color_legend_1.colorLegend,
    sizeLegend: size_legend_1.sizeLegend,
    axis: common_axis_1.commonAxis,
    axisBand: band_axis_1.axisBand,
    axisLinear: linear_axis_1.axisLinear,
    axisX: cartesian_axis_1.axisX,
    axisY: cartesian_axis_1.axisY,
    axisZ: cartesian_axis_1.axisZ,
    axisAngle: polar_axis_1.axisAngle,
    axisRadius: polar_axis_1.axisRadius,
    markLine: mark_line_1.markLine,
    markArea: mark_area_1.markArea,
    markPoint: mark_point_1.markPoint,
    tooltip: tooltip_1.tooltip,
    dataZoom: data_zoom_1.dataZoom,
    crosshair: crosshair_1.crosshair,
    player: player_1.player,
    brush: brush_1.brush,
    indicator: indicator_1.indicator,
    title: title_1.title,
    mapLabel: map_label_1.mapLabel,
    poptip: poptip_1.poptip,
    totalLabel: total_label_1.totalLabel,
    scrollBar: scroll_bar_1.scrollBar
};
