// @visactor/vchart/cjs/theme/builtin/common/component/mark-area.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.markArea = void 0;

const constants_1 = require("../constants");

exports.markArea = {
    area: {
        style: {
            fill: {
                type: "palette",
                key: "axisDomainColor",
                a: .25
            }
        }
    },
    label: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l4FontSize,
            fontWeight: "normal",
            fontStyle: "normal",
            fill: {
                type: "palette",
                key: "primaryFontColor"
            }
        },
        labelBackground: {
            padding: {
                top: 2,
                bottom: 2,
                right: 4,
                left: 4
            },
            style: {
                cornerRadius: 3,
                fill: {
                    type: "palette",
                    key: "markLabelBackgroundColor"
                }
            }
        }
    }
};
