// @visactor/vchart/esm/series/link/tooltip-helper.js
import { BaseSeriesTooltipHelper } from "../base/tooltip-helper";

import { TimeUtil } from "../../component/axis/cartesian/util/time";

export class LinkSeriesTooltipHelper extends BaseSeriesTooltipHelper {
    getDefaultTooltipPattern(activeType) {
        return "mark" === activeType ? {
            visible: !0,
            activeType: activeType,
            title: {
                key: "link info",
                value: "link info"
            },
            content: [ {
                hasShape: !0,
                shapeType: "square",
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                key: "time",
                value: datum => TimeUtil.getInstance().timeFormat("%Y%m%d %H:%M", datum.from.split("_")[1])
            }, {
                hasShape: !0,
                shapeType: "square",
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                key: "type",
                value: datum => datum.action_type
            }, {
                hasShape: !0,
                shapeType: "square",
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                key: "from",
                value: datum => datum.from
            }, {
                hasShape: !0,
                shapeType: "square",
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                key: "to",
                value: datum => datum.to
            } ]
        } : null;
    }
}
