// @visactor/vchart/esm/series/range-column/tooltip-helper.js
import { BaseSeriesTooltipHelper } from "../base/tooltip-helper";

import { Direction } from "../../typings/space";

export class RangeColumnSeriesTooltipHelper extends BaseSeriesTooltipHelper {
    getDefaultTooltipPattern(activeType) {
        return {
            visible: !0,
            activeType: activeType,
            title: {
                key: void 0,
                value: this.titleValueCallback,
                hasShape: !1
            },
            content: [ {
                key: this.contentKeyCallback,
                value: datum => this.series.getSpec().direction === Direction.horizontal ? datum[this.series.getSpec().xField[0]] + "-" + datum[this.series.getSpec().xField[1]] : datum[this.series.getSpec().yField[0]] + "-" + datum[this.series.getSpec().yField[1]],
                hasShape: !0,
                shapeType: this.contentShapeTypeCallback,
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                shapeHollow: !1
            } ]
        };
    }
}
