// @visactor/vchart/cjs/constant/funnel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.FUNNEL_MIN_SIZE = exports.FUNNEL_MAX_SIZE = exports.FUNNEL_LABEL_SPACE_WIDTH = exports.FUNNEL_LABEL_LINE_LENGTH = exports.FUNNEL_TRANSFORM_LEVEL = exports.FUNNEL_NEXT_VALUE = exports.FUNNEL_CURRENT_VALUE = exports.FUNNEL_LAST_VALUE = exports.FUNNEL_NEXT_VALUE_RATIO = exports.FUNNEL_LAST_VALUE_RATIO = exports.FUNNEL_VALUE_RATIO = exports.FUNNEL_HEIGHT_RATIO = exports.FUNNEL_REACH_RATIO = exports.FUNNEL_TRANSFORM_RATIO = void 0;

const base_1 = require("./base");

exports.FUNNEL_TRANSFORM_RATIO = `${base_1.PREFIX}_FUNNEL_TRANSFORM_RATIO`, exports.FUNNEL_REACH_RATIO = `${base_1.PREFIX}_FUNNEL_REACH_RATIO`, 
exports.FUNNEL_HEIGHT_RATIO = `${base_1.PREFIX}_FUNNEL_HEIGHT_RATIO`, exports.FUNNEL_VALUE_RATIO = `${base_1.PREFIX}_FUNNEL_VALUE_RATIO`, 
exports.FUNNEL_LAST_VALUE_RATIO = `${base_1.PREFIX}_FUNNEL_LAST_VALUE_RATIO`, exports.FUNNEL_NEXT_VALUE_RATIO = `${base_1.PREFIX}_FUNNEL_NEXT_VALUE_RATIO`, 
exports.FUNNEL_LAST_VALUE = `${base_1.PREFIX}_FUNNEL_LAST_VALUE`, exports.FUNNEL_CURRENT_VALUE = `${base_1.PREFIX}_FUNNEL_CURRENT_VALUE`, 
exports.FUNNEL_NEXT_VALUE = `${base_1.PREFIX}_FUNNEL_NEXT_VALUE`, exports.FUNNEL_TRANSFORM_LEVEL = `${base_1.PREFIX}_FUNNEL_TRANSFORM_LEVEL`, 
exports.FUNNEL_LABEL_LINE_LENGTH = 20, exports.FUNNEL_LABEL_SPACE_WIDTH = 5, exports.FUNNEL_MAX_SIZE = "80%", 
exports.FUNNEL_MIN_SIZE = 0;
