// @visactor/vchart/cjs/constant/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.DEFAULT_GRADIENT_CONFIG = exports.DEFAULT_CONICAL_GRADIENT_CONFIG = exports.DEFAULT_RADIAL_GRADIENT_CONFIG = exports.DEFAULT_LINEAR_GRADIENT_CONFIG = exports.GradientType = exports.LayoutLevel = exports.LayoutZIndex = exports.SEGMENT_FIELD_END = exports.SEGMENT_FIELD_START = exports.STACK_FIELD_TOTAL_TOP = exports.STACK_FIELD_TOTAL_PERCENT = exports.STACK_FIELD_TOTAL = exports.STACK_FIELD_END_OffsetSilhouette = exports.STACK_FIELD_START_OffsetSilhouette = exports.STACK_FIELD_END_PERCENT = exports.STACK_FIELD_START_PERCENT = exports.STACK_FIELD_END = exports.STACK_FIELD_START = exports.AttributeLevel = exports.DEFAULT_SERIES_STYLE_NAME = exports.DEFAULT_DATA_SERIES_FIELD = exports.DEFAULT_DATA_KEY = exports.DEFAULT_DATA_INDEX = exports.DEFAULT_MEASURE_CANVAS_ID = void 0;

const base_1 = require("./base");

var AttributeLevel, LayoutZIndex, LayoutLevel;

exports.DEFAULT_MEASURE_CANVAS_ID = `${base_1.PREFIX}_MEASURE_CANVAS_ID`, exports.DEFAULT_DATA_INDEX = `${base_1.PREFIX}_DEFAULT_DATA_INDEX`, 
exports.DEFAULT_DATA_KEY = `${base_1.PREFIX}_DEFAULT_DATA_KEY`, exports.DEFAULT_DATA_SERIES_FIELD = `${base_1.PREFIX}_DEFAULT_DATA_SERIES_FIELD`, 
exports.DEFAULT_SERIES_STYLE_NAME = `${base_1.PREFIX}_DEFAULT_SERIES_STYLE_NAME`, 
function(AttributeLevel) {
    AttributeLevel[AttributeLevel.Default = 0] = "Default", AttributeLevel[AttributeLevel.Theme = 1] = "Theme", 
    AttributeLevel[AttributeLevel.Chart = 2] = "Chart", AttributeLevel[AttributeLevel.Base_Series = 3] = "Base_Series", 
    AttributeLevel[AttributeLevel.Series = 4] = "Series", AttributeLevel[AttributeLevel.Mark = 5] = "Mark", 
    AttributeLevel[AttributeLevel.User_Chart = 6] = "User_Chart", AttributeLevel[AttributeLevel.User_Series = 7] = "User_Series", 
    AttributeLevel[AttributeLevel.User_Mark = 8] = "User_Mark", AttributeLevel[AttributeLevel.Built_In = 99] = "Built_In";
}(AttributeLevel = exports.AttributeLevel || (exports.AttributeLevel = {})), exports.STACK_FIELD_START = `${base_1.PREFIX}_STACK_START`, 
exports.STACK_FIELD_END = `${base_1.PREFIX}_STACK_END`, exports.STACK_FIELD_START_PERCENT = `${base_1.PREFIX}_STACK_START_PERCENT`, 
exports.STACK_FIELD_END_PERCENT = `${base_1.PREFIX}_STACK_END_PERCENT`, exports.STACK_FIELD_START_OffsetSilhouette = `${base_1.PREFIX}_STACK_START_OffsetSilhouette`, 
exports.STACK_FIELD_END_OffsetSilhouette = `${base_1.PREFIX}_STACK_END_OffsetSilhouette`, 
exports.STACK_FIELD_TOTAL = `${base_1.PREFIX}_STACK_TOTAL`, exports.STACK_FIELD_TOTAL_PERCENT = `${base_1.PREFIX}_STACK_TOTAL_PERCENT`, 
exports.STACK_FIELD_TOTAL_TOP = `${base_1.PREFIX}_STACK_TOTAL_TOP`, exports.SEGMENT_FIELD_START = `${base_1.PREFIX}_SEGMENT_START`, 
exports.SEGMENT_FIELD_END = `${base_1.PREFIX}_SEGMENT_END`, function(LayoutZIndex) {
    LayoutZIndex[LayoutZIndex.Axis_Grid = 50] = "Axis_Grid", LayoutZIndex[LayoutZIndex.CrossHair_Grid = 100] = "CrossHair_Grid", 
    LayoutZIndex[LayoutZIndex.Region = 450] = "Region", LayoutZIndex[LayoutZIndex.Mark = 300] = "Mark", 
    LayoutZIndex[LayoutZIndex.Node = 400] = "Node", LayoutZIndex[LayoutZIndex.Axis = 100] = "Axis", 
    LayoutZIndex[LayoutZIndex.MarkLine = 500] = "MarkLine", LayoutZIndex[LayoutZIndex.MarkArea = 100] = "MarkArea", 
    LayoutZIndex[LayoutZIndex.MarkPoint = 500] = "MarkPoint", LayoutZIndex[LayoutZIndex.DataZoom = 500] = "DataZoom", 
    LayoutZIndex[LayoutZIndex.ScrollBar = 500] = "ScrollBar", LayoutZIndex[LayoutZIndex.Player = 500] = "Player", 
    LayoutZIndex[LayoutZIndex.Legend = 500] = "Legend", LayoutZIndex[LayoutZIndex.CrossHair = 500] = "CrossHair", 
    LayoutZIndex[LayoutZIndex.Indicator = 500] = "Indicator", LayoutZIndex[LayoutZIndex.Title = 500] = "Title", 
    LayoutZIndex[LayoutZIndex.Label = 500] = "Label", LayoutZIndex[LayoutZIndex.Brush = 500] = "Brush", 
    LayoutZIndex[LayoutZIndex.CustomMark = 500] = "CustomMark", LayoutZIndex[LayoutZIndex.Interaction = 700] = "Interaction";
}(LayoutZIndex = exports.LayoutZIndex || (exports.LayoutZIndex = {})), function(LayoutLevel) {
    LayoutLevel[LayoutLevel.Indicator = 10] = "Indicator", LayoutLevel[LayoutLevel.Region = 20] = "Region", 
    LayoutLevel[LayoutLevel.Axis = 30] = "Axis", LayoutLevel[LayoutLevel.DataZoom = 40] = "DataZoom", 
    LayoutLevel[LayoutLevel.Player = 40] = "Player", LayoutLevel[LayoutLevel.ScrollBar = 40] = "ScrollBar", 
    LayoutLevel[LayoutLevel.Legend = 50] = "Legend", LayoutLevel[LayoutLevel.Title = 70] = "Title", 
    LayoutLevel[LayoutLevel.CustomMark = 70] = "CustomMark";
}(LayoutLevel = exports.LayoutLevel || (exports.LayoutLevel = {})), exports.GradientType = [ "linear", "radial", "conical" ], 
exports.DEFAULT_LINEAR_GRADIENT_CONFIG = {
    x0: 0,
    y0: 0,
    x1: 1,
    y1: 1
}, exports.DEFAULT_RADIAL_GRADIENT_CONFIG = {
    x0: 0,
    y0: 0,
    x1: 1,
    y1: 1,
    r0: 0,
    r1: 1
}, exports.DEFAULT_CONICAL_GRADIENT_CONFIG = {
    x: .5,
    y: .5,
    startAngle: 0,
    endAngle: 2 * Math.PI
}, exports.DEFAULT_GRADIENT_CONFIG = {
    linear: exports.DEFAULT_LINEAR_GRADIENT_CONFIG,
    radial: exports.DEFAULT_RADIAL_GRADIENT_CONFIG,
    conical: exports.DEFAULT_CONICAL_GRADIENT_CONFIG
}, __exportStar(require("./base"), exports), __exportStar(require("./label"), exports), 
__exportStar(require("./polar"), exports), __exportStar(require("./layout"), exports), 
__exportStar(require("./event"), exports), __exportStar(require("./waterfall"), exports), 
__exportStar(require("./correlation"), exports);
