// @visactor/vchart/cjs/constant/polar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.POLAR_DEFAULT_RADIUS = exports.POLAR_END_ANGLE = exports.POLAR_START_ANGLE = exports.POLAR_END_RADIAN = exports.POLAR_START_RADIAN = exports.ARC_RADIAN = exports.ARC_QUADRANT = exports.ARC_MIDDLE_ANGLE = exports.ARC_K = exports.ARC_END_ANGLE = exports.ARC_START_ANGLE = exports.ARC_RATIO = exports.ARC_TRANSFORM_VALUE = void 0;

const base_1 = require("./base");

exports.ARC_TRANSFORM_VALUE = `${base_1.PREFIX}_ARC_TRANSFORM_VALUE`, exports.ARC_RATIO = `${base_1.PREFIX}_ARC_RATIO`, 
exports.ARC_START_ANGLE = `${base_1.PREFIX}_ARC_START_ANGLE`, exports.ARC_END_ANGLE = `${base_1.PREFIX}_ARC_END_ANGLE`, 
exports.ARC_K = `${base_1.PREFIX}_ARC_K`, exports.ARC_MIDDLE_ANGLE = `${base_1.PREFIX}_ARC_MIDDLE_ANGLE`, 
exports.ARC_QUADRANT = `${base_1.PREFIX}_ARC_QUADRANT`, exports.ARC_RADIAN = `${base_1.PREFIX}_ARC_RADIAN`, 
exports.POLAR_START_RADIAN = -Math.PI / 2, exports.POLAR_END_RADIAN = 3 * Math.PI / 2, 
exports.POLAR_START_ANGLE = -90, exports.POLAR_END_ANGLE = 270, exports.POLAR_DEFAULT_RADIUS = .6;
