// @visactor/vchart/cjs/util/env.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.isMiniAppLikeMode = exports.isMobileLikeMode = exports.isPcLikeMode = exports.isTrueBrowser = exports.domDocument = exports.isBrowser = void 0;

const vrender_core_1 = require("@visactor/vrender-core");

function isTrueBrowser(mode) {
    return ("desktop-browser" === mode || "mobile-browser" === mode) && exports.isBrowser;
}

function isPcLikeMode(mode) {
    return "desktop-miniApp" === mode || "desktop-browser" === mode;
}

function isMobileLikeMode(mode) {
    return isMiniAppLikeMode(mode) || "mobile-browser" === mode;
}

function isMiniAppLikeMode(mode) {
    return mode.includes("miniApp") || "lynx" === mode || "wx" === mode;
}

exports.isBrowser = (0, vrender_core_1.isBrowserEnv)(), exports.domDocument = exports.isBrowser && globalThis ? globalThis.document : void 0, 
exports.isTrueBrowser = isTrueBrowser, exports.isPcLikeMode = isPcLikeMode, exports.isMobileLikeMode = isMobileLikeMode, 
exports.isMiniAppLikeMode = isMiniAppLikeMode;
