// @visactor/vdataset/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.readCSVTopNLine = exports.DataView = exports.DataSet = exports.dataViewParser = exports.treeParser = exports.byteJSONParser = exports.topoJSONParser = exports.geoJSONParser = exports.geoBufParser = exports.tsvParser = exports.csvParser = exports.dsvParser = exports.fields = exports.fold = exports.map = exports.statistics = exports.mercator = exports.simplify = exports.dissolve = exports.projection = exports.filter = exports.pointToHexbin = void 0;

var hexagon_1 = require("./transform/hexagon");

Object.defineProperty(exports, "pointToHexbin", {
    enumerable: !0,
    get: function() {
        return hexagon_1.pointToHexbin;
    }
});

var filter_1 = require("./transform/filter");

Object.defineProperty(exports, "filter", {
    enumerable: !0,
    get: function() {
        return filter_1.filter;
    }
});

var projection_1 = require("./transform/geo/projection");

Object.defineProperty(exports, "projection", {
    enumerable: !0,
    get: function() {
        return projection_1.projection;
    }
});

var dissolve_1 = require("./transform/geo/dissolve");

Object.defineProperty(exports, "dissolve", {
    enumerable: !0,
    get: function() {
        return dissolve_1.dissolve;
    }
});

var simplify_1 = require("./transform/geo/simplify");

Object.defineProperty(exports, "simplify", {
    enumerable: !0,
    get: function() {
        return simplify_1.simplify;
    }
});

var mercator_1 = require("./transform/geo/mercator");

Object.defineProperty(exports, "mercator", {
    enumerable: !0,
    get: function() {
        return mercator_1.mercator;
    }
});

var statistics_1 = require("./transform/statistics");

Object.defineProperty(exports, "statistics", {
    enumerable: !0,
    get: function() {
        return statistics_1.statistics;
    }
});

var map_1 = require("./transform/map");

Object.defineProperty(exports, "map", {
    enumerable: !0,
    get: function() {
        return map_1.map;
    }
});

var fold_1 = require("./transform/fold");

Object.defineProperty(exports, "fold", {
    enumerable: !0,
    get: function() {
        return fold_1.fold;
    }
});

var fields_1 = require("./transform/fields");

Object.defineProperty(exports, "fields", {
    enumerable: !0,
    get: function() {
        return fields_1.fields;
    }
}), __exportStar(require("./transform/index"), exports);

var dsv_1 = require("./parser/dsv");

Object.defineProperty(exports, "dsvParser", {
    enumerable: !0,
    get: function() {
        return dsv_1.dsvParser;
    }
}), Object.defineProperty(exports, "csvParser", {
    enumerable: !0,
    get: function() {
        return dsv_1.csvParser;
    }
}), Object.defineProperty(exports, "tsvParser", {
    enumerable: !0,
    get: function() {
        return dsv_1.tsvParser;
    }
});

var geobuf_1 = require("./parser/geo/geobuf");

Object.defineProperty(exports, "geoBufParser", {
    enumerable: !0,
    get: function() {
        return geobuf_1.geoBufParser;
    }
});

var geojson_1 = require("./parser/geo/geojson");

Object.defineProperty(exports, "geoJSONParser", {
    enumerable: !0,
    get: function() {
        return geojson_1.geoJSONParser;
    }
});

var topojson_1 = require("./parser/geo/topojson");

Object.defineProperty(exports, "topoJSONParser", {
    enumerable: !0,
    get: function() {
        return topojson_1.topoJSONParser;
    }
});

var bytejson_1 = require("./parser/bytejson");

Object.defineProperty(exports, "byteJSONParser", {
    enumerable: !0,
    get: function() {
        return bytejson_1.byteJSONParser;
    }
});

var tree_1 = require("./parser/tree");

Object.defineProperty(exports, "treeParser", {
    enumerable: !0,
    get: function() {
        return tree_1.treeParser;
    }
});

var data_view_1 = require("./parser/data-view");

Object.defineProperty(exports, "dataViewParser", {
    enumerable: !0,
    get: function() {
        return data_view_1.dataViewParser;
    }
});

var data_set_1 = require("./data-set");

Object.defineProperty(exports, "DataSet", {
    enumerable: !0,
    get: function() {
        return data_set_1.DataSet;
    }
});

var data_view_2 = require("./data-view");

Object.defineProperty(exports, "DataView", {
    enumerable: !0,
    get: function() {
        return data_view_2.DataView;
    }
});

var csv_1 = require("./utils/csv");

Object.defineProperty(exports, "readCSVTopNLine", {
    enumerable: !0,
    get: function() {
        return csv_1.readCSVTopNLine;
    }
});
