// @visactor/vchart/esm/chart/pie/3d/pie-3d.js
import { registerPie3dSeries } from "../../../series/pie/3d/pie-3d";

import { SeriesTypeEnum } from "../../../series/interface/type";

import { ChartTypeEnum } from "../../interface/type";

import { BasePieChart } from "../base/base";

import { Factory } from "../../../core/factory";

import { BasePieChartSpecTransformer } from "../base";

export class Pie3dChartSpecTransformer extends BasePieChartSpecTransformer {
    transformSpec(spec) {
        super.transformSpec(spec), spec.series.forEach((s => {
            "pie3d" === s.type && (s.angle3d = spec.angle3d);
        }));
    }
}

export class Pie3dChart extends BasePieChart {
    constructor() {
        super(...arguments), this.transformerConstructor = Pie3dChartSpecTransformer, this.type = ChartTypeEnum.pie3d, 
        this.seriesType = SeriesTypeEnum.pie3d;
    }
}

Pie3dChart.type = ChartTypeEnum.pie3d, Pie3dChart.seriesType = SeriesTypeEnum.pie3d, 
Pie3dChart.view = "singleDefault", Pie3dChart.transformerConstructor = Pie3dChartSpecTransformer;

export const registerPie3dChart = () => {
    registerPie3dSeries(), Factory.registerChart(Pie3dChart.type, Pie3dChart);
};
