// @visactor/vchart/cjs/series/progress/circular/circular.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerCircularProgressSeries = exports.CircularProgressSeries = void 0;

const vutils_1 = require("@visactor/vutils"), type_1 = require("../../interface/type"), utils_1 = require("../../../animation/utils"), progress_like_1 = require("../../polar/progress-like/progress-like"), arc_1 = require("../../../mark/arc"), progress_arc_1 = require("../../../mark/progress-arc"), constant_1 = require("./constant"), constant_2 = require("../../../constant"), factory_1 = require("../../../core/factory"), progress_like_2 = require("../../polar/progress-like"), config_1 = require("../../../animation/config");

class CircularProgressSeries extends progress_like_1.ProgressLikeSeries {
    constructor() {
        super(...arguments), this.type = type_1.SeriesTypeEnum.circularProgress, this._progressMark = null, 
        this._trackMark = null, this._getRadiusValueStart = datum => {
            if (this.getGroupFields().length > 1) {
                const value = this.radiusAxisHelper.dataToPosition(this.getDatumPositionValues(datum, this.getGroupFields()));
                if ((0, vutils_1.isValidNumber)(value)) return value;
            }
            return this.radiusAxisHelper.dataToPosition([ datum[this._radiusField[0]] ]);
        }, this._getRadiusValueEnd = datum => {
            var _a, _b;
            if (this.getGroupFields().length > 1) {
                const value = this.radiusAxisHelper.dataToPosition(this.getDatumPositionValues(datum, this.getGroupFields())) + (null === (_b = (_a = this.radiusAxisHelper).getBandwidth) || void 0 === _b ? void 0 : _b.call(_a, this._groups ? this._groups.fields.length - 1 : 0));
                if ((0, vutils_1.isValidNumber)(value)) return value;
            }
            return this.radiusAxisHelper.dataToPosition([ datum[this._radiusField[0]] ]) + this.radiusAxisHelper.getScale(0).step();
        };
    }
    getStackGroupFields() {
        return this.getGroupFields();
    }
    getGroupFields() {
        return this._radiusField;
    }
    initMark() {
        super.initMark(), this._initTrackMark(), this._initProgressMark();
    }
    initMarkStyle() {
        super.initMarkStyle(), this._initTrackMarkStyle(), this._initProgressMarkStyle();
    }
    _initProgressMark() {
        var _a;
        return this._progressMark = this._createMark(CircularProgressSeries.mark.progress, {
            parent: this._arcGroupMark,
            isSeriesMark: !0,
            customShape: null === (_a = this._spec.progress) || void 0 === _a ? void 0 : _a.customShape
        }), this._progressMark;
    }
    _initProgressMarkStyle() {
        var _a;
        const progressMark = this._progressMark;
        progressMark && (this.setMarkStyle(progressMark, {
            x: () => this.angleAxisHelper.center().x,
            y: () => this.angleAxisHelper.center().y,
            startAngle: this._getAngleValueStart,
            endAngle: this._getAngleValueEnd,
            innerRadius: this._getRadiusValueStart,
            outerRadius: this._getRadiusValueEnd,
            cap: null !== (_a = this._spec.roundCap) && void 0 !== _a && _a,
            boundsMode: "imprecise",
            cornerRadius: this._spec.cornerRadius,
            fill: this.getColorAttribute(),
            zIndex: 200,
            forceShowCap: !0
        }, "normal", constant_2.AttributeLevel.Series), this._trigger.registerMark(progressMark));
    }
    initTooltip() {
        super.initTooltip(), this._progressMark && this._tooltipHelper.activeTriggerSet.mark.add(this._progressMark);
    }
    _initTrackMark() {
        var _a;
        return this._trackMark = this._createMark(CircularProgressSeries.mark.track, {
            parent: this._arcGroupMark,
            customShape: null === (_a = this._spec.track) || void 0 === _a ? void 0 : _a.customShape
        }), this._trackMark;
    }
    _initTrackMarkStyle() {
        const trackMark = this._trackMark;
        trackMark && (this.setMarkStyle(trackMark, {
            visible: datum => {
                const range = this.angleAxisHelper.getScale(0).range(), min = Math.min(range[0], range[range.length - 1]), startValue = this._getAngleValueStartWithoutMask(datum);
                return Math.abs(startValue - min) <= 1e-14;
            },
            x: () => this.angleAxisHelper.center().x,
            y: () => this.angleAxisHelper.center().y,
            startAngle: () => {
                const fieldName = this._stack ? constant_2.STACK_FIELD_START : this._angleField[0], domain = this.angleAxisHelper.getScale(0).domain();
                return this._getAngleValueStart({
                    [fieldName]: domain[0]
                });
            },
            endAngle: () => {
                const fieldName = this._stack ? constant_2.STACK_FIELD_END : this._angleField[0], domain = this.angleAxisHelper.getScale(0).domain();
                return this._getAngleValueEnd({
                    [fieldName]: domain[domain.length - 1]
                });
            },
            innerRadius: this._getRadiusValueStart,
            outerRadius: this._getRadiusValueEnd,
            cornerRadius: this._spec.cornerRadius,
            fill: this.getColorAttribute(),
            zIndex: 100
        }, "normal", constant_2.AttributeLevel.Series), this._trigger.registerMark(trackMark));
    }
    initAnimation() {
        var _a, _b, _c, _d;
        const appearPreset = null === (_b = null === (_a = this._spec) || void 0 === _a ? void 0 : _a.animationAppear) || void 0 === _b ? void 0 : _b.preset;
        this._progressMark.setAnimationConfig((0, utils_1.animationConfig)(null === (_c = factory_1.Factory.getAnimationInKey("circularProgress")) || void 0 === _c ? void 0 : _c({
            startAngle: this._startAngle
        }, appearPreset), (0, utils_1.userAnimationConfig)(type_1.SeriesMarkNameEnum.progress, this._spec, this._markAttributeContext))), 
        this._trackMark.setAnimationConfig((0, utils_1.animationConfig)(null === (_d = factory_1.Factory.getAnimationInKey("fadeInOut")) || void 0 === _d ? void 0 : _d(), (0, 
        utils_1.userAnimationConfig)(type_1.SeriesMarkNameEnum.track, this._spec, this._markAttributeContext)));
    }
    getActiveMarks() {
        return [ this._progressMark ];
    }
}

exports.CircularProgressSeries = CircularProgressSeries, CircularProgressSeries.type = type_1.SeriesTypeEnum.circularProgress, 
CircularProgressSeries.mark = constant_1.circularProgressSeriesMark;

const registerCircularProgressSeries = () => {
    (0, progress_arc_1.registerProgressArcMark)(), (0, arc_1.registerArcMark)(), (0, 
    progress_like_2.registerProgressLikeAnimation)(), (0, config_1.registerFadeInOutAnimation)(), 
    factory_1.Factory.registerSeries(CircularProgressSeries.type, CircularProgressSeries);
};

exports.registerCircularProgressSeries = registerCircularProgressSeries;
