// @visactor/vchart-semi-theme/esm/common/token-map.js
export const paletteTokenMap = {
    backgroundColor: "--semi-color-bg-0",
    borderColor: "--semi-color-border",
    hoverBackgroundColor: "--semi-color-fill-0",
    sliderRailColor: "--semi-color-fill-0",
    sliderHandleColor: "--semi-white",
    sliderTrackColor: "--semi-color-primary",
    popupBackgroundColor: "--semi-color-bg-3",
    primaryFontColor: "--semi-color-text-0",
    secondaryFontColor: "--semi-color-text-1",
    tertiaryFontColor: "--semi-color-text-2",
    axisLabelFontColor: "--semi-color-text-0",
    disableFontColor: "--semi-color-disabled-text",
    axisMarkerFontColor: "--semi-color-bg-0",
    axisGridColor: "--semi-color-border",
    axisDomainColor: {
        light: "--semi-grey-9",
        dark: "--semi-color-border"
    },
    dataZoomHandleStrokeColor: {
        light: "--semi-color-fill-2"
    },
    dataZoomChartColor: "--semi-color-fill-1",
    playerControllerColor: "--semi-color-primary",
    axisMarkerBackgroundColor: "--semi-color-text-0",
    markLabelBackgroundColor: "--semi-color-border",
    markLineStrokeColor: "--semi-color-text-1",
    dangerColor: "--semi-color-danger",
    warningColor: "--semi-color-warning",
    successColor: "--semi-color-success",
    infoColor: "--semi-color-info"
};

export const dataSchemeTokenMap = [ {
    maxDomainLength: 10,
    scheme: [ "--semi-color-data-0", "--semi-color-data-2", "--semi-color-data-4", "--semi-color-data-6", "--semi-color-data-8", "--semi-color-data-10", "--semi-color-data-12", "--semi-color-data-14", "--semi-color-data-16", "--semi-color-data-18" ]
}, {
    scheme: [ "--semi-color-data-0", "--semi-color-data-1", "--semi-color-data-2", "--semi-color-data-3", "--semi-color-data-4", "--semi-color-data-5", "--semi-color-data-6", "--semi-color-data-7", "--semi-color-data-8", "--semi-color-data-9", "--semi-color-data-10", "--semi-color-data-11", "--semi-color-data-12", "--semi-color-data-13", "--semi-color-data-14", "--semi-color-data-15", "--semi-color-data-16", "--semi-color-data-17", "--semi-color-data-18", "--semi-color-data-19" ]
} ];
