// @visactor/vchart/cjs/chart/word-cloud/3d/word-cloud-3d-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.WordCloud3dChartSpecTransformer = void 0;

const word_cloud_base_transformer_1 = require("../base/word-cloud-base-transformer");

class WordCloud3dChartSpecTransformer extends word_cloud_base_transformer_1.BaseWordCloudChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            nameField: spec.nameField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            fontFamilyField: spec.fontFamilyField,
            fontWeightField: spec.fontWeightField,
            fontStyleField: spec.fontStyleField,
            colorHexField: spec.colorHexField,
            colorMode: spec.colorMode,
            colorList: spec.colorList,
            rotateAngles: spec.rotateAngles,
            fontWeightRange: spec.fontWeightRange,
            fontSizeRange: spec.fontSizeRange,
            depth_3d: spec.depth_3d,
            maskShape: spec.maskShape,
            keepAspect: spec.keepAspect,
            random: spec.random,
            wordCloudConfig: spec.wordCloudConfig,
            wordCloudShapeConfig: spec.wordCloudShapeConfig,
            word: spec.word,
            fillingWord: spec.fillingWord
        }), seriesType = this.seriesType;
        return seriesType && (series.type = seriesType, series[seriesType] = spec[seriesType]), 
        series;
    }
}

exports.WordCloud3dChartSpecTransformer = WordCloud3dChartSpecTransformer;
