// @visactor/vchart/esm/theme/builtin/common/component/legend/continuous.js
import { THEME_CONSTANTS } from "../../constants";

export const DEFAULT_CONTINUOUS_LEGEND_THEME = {
    orient: "right",
    position: "middle",
    padding: [ 16, 24 ],
    title: {
        visible: !1,
        padding: 0,
        textStyle: {
            fontSize: THEME_CONSTANTS.l5FontSize,
            lineHeight: THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "primaryFontColor"
            }
        },
        space: 12
    },
    handler: {
        visible: !0
    },
    startText: {
        style: {
            fontSize: THEME_CONSTANTS.l5FontSize,
            lineHeight: THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    },
    endText: {
        style: {
            fontSize: THEME_CONSTANTS.l5FontSize,
            lineHeight: THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    },
    handlerText: {
        style: {
            fontSize: THEME_CONSTANTS.l5FontSize,
            lineHeight: THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    }
};
