// @visactor/vchart/cjs/interaction/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.getDefaultInteractionConfigByMode = void 0;

const common_1 = require("../typings/spec/common"), util_1 = require("../util");

function getDefaultInteractionConfigByMode(mode) {
    return mode === common_1.RenderModeEnum["desktop-browser"] || mode === common_1.RenderModeEnum["desktop-miniApp"] ? {
        hover: {
            enable: !0,
            trigger: "pointermove",
            triggerOff: "pointerleave"
        },
        select: {
            enable: !0,
            trigger: "pointertap"
        }
    } : (0, util_1.isMobileLikeMode)(mode) || (0, util_1.isMiniAppLikeMode)(mode) ? {
        hover: {
            enable: !0,
            trigger: [ "pointerdown", "pointermove" ],
            triggerOff: "pointerleave"
        },
        select: {
            enable: !0,
            trigger: "tap"
        }
    } : null;
}

exports.getDefaultInteractionConfigByMode = getDefaultInteractionConfigByMode;
