// @visactor/vchart/cjs/theme/builtin/common/component/legend/continuous.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.DEFAULT_CONTINUOUS_LEGEND_THEME = void 0;

const constants_1 = require("../../constants");

exports.DEFAULT_CONTINUOUS_LEGEND_THEME = {
    orient: "right",
    position: "middle",
    padding: [ 16, 24 ],
    title: {
        visible: !1,
        padding: 0,
        textStyle: {
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "primaryFontColor"
            }
        },
        space: 12
    },
    handler: {
        visible: !0
    },
    startText: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    },
    endText: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    },
    handlerText: {
        style: {
            fontSize: constants_1.THEME_CONSTANTS.l5FontSize,
            lineHeight: constants_1.THEME_CONSTANTS.l5LineHeight,
            fontWeight: "normal",
            fill: {
                type: "palette",
                key: "secondaryFontColor",
                default: "#89909d"
            }
        },
        space: 6
    }
};
