// @visactor/vchart/cjs/chart/line/line-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.LineChartSpecTransformer = void 0;

const cartesian_1 = require("../cartesian"), util_1 = require("../util");

class LineChartSpecTransformer extends cartesian_1.CartesianChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        var _a;
        return Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            point: spec.point,
            line: spec.line,
            seriesMark: null !== (_a = spec.seriesMark) && void 0 !== _a ? _a : "line",
            activePoint: spec.activePoint,
            sampling: spec.sampling,
            samplingFactor: spec.samplingFactor,
            pointDis: spec.pointDis,
            pointDisMul: spec.pointDisMul,
            markOverlap: spec.markOverlap,
            lineLabel: spec.lineLabel
        });
    }
    transformSpec(spec) {
        super.transformSpec(spec), (0, util_1.setDefaultCrosshairForCartesianChart)(spec);
    }
}

exports.LineChartSpecTransformer = LineChartSpecTransformer;
