// @visactor/vchart/cjs/component/marker/base-marker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.BaseMarker = void 0;

const vutils_1 = require("@visactor/vutils"), base_component_1 = require("../base/base-component"), space_1 = require("../../util/space"), utils_1 = require("./utils"), util_1 = require("../../util");

class BaseMarker extends base_component_1.BaseComponent {
    constructor() {
        super(...arguments), this.layoutType = "none", this._layoutOffsetX = 0, this._layoutOffsetY = 0;
    }
    getRelativeSeries() {
        return this._relativeSeries;
    }
    created() {
        super.created(), this.initEvent(), this._bindSeries(), this._initDataView();
    }
    _getAllRelativeSeries() {
        return {
            getRelativeSeries: () => this._relativeSeries,
            getStartRelativeSeries: () => this._startRelativeSeries,
            getEndRelativeSeries: () => this._endRelativeSeries
        };
    }
    _getFieldInfoFromSpec(dim, spec, relativeSeries) {
        const field = "x" === dim ? relativeSeries.getSpec().xField : relativeSeries.getSpec().yField;
        return (0, vutils_1.isString)(spec) && (0, utils_1.isAggrSpec)(spec) ? {
            field: field,
            aggrType: spec
        } : spec;
    }
    _processSpecX(specX) {
        const relativeSeries = this._relativeSeries;
        return Object.assign({
            x: this._getFieldInfoFromSpec("x", specX, relativeSeries)
        }, this._getAllRelativeSeries());
    }
    _processSpecY(specY) {
        const relativeSeries = this._relativeSeries;
        return Object.assign({
            y: this._getFieldInfoFromSpec("y", specY, relativeSeries)
        }, this._getAllRelativeSeries());
    }
    _processSpecXY(specX, specY) {
        const relativeSeries = this._relativeSeries;
        return Object.assign({
            x: this._getFieldInfoFromSpec("x", specX, relativeSeries),
            y: this._getFieldInfoFromSpec("y", specY, relativeSeries)
        }, this._getAllRelativeSeries());
    }
    _processSpecCoo(spec) {
        var _a;
        return (null !== (_a = spec.coordinates) && void 0 !== _a ? _a : (0, vutils_1.array)(spec.coordinate)).map((coordinate => {
            const refRelativeSeries = this._getSeriesByIdOrIndex(coordinate.refRelativeSeriesId, coordinate.refRelativeSeriesIndex), {xField: xField, yField: yField} = refRelativeSeries.getSpec(), {xFieldDim: xFieldDim, xFieldIndex: xFieldIndex, yFieldDim: yFieldDim, yFieldIndex: yFieldIndex} = coordinate;
            let bindXField = xField;
            (0, vutils_1.isValid)(xFieldIndex) && (bindXField = (0, vutils_1.array)(xField)[xFieldIndex]), 
            xFieldDim && (0, vutils_1.array)(xField).includes(xFieldDim) && (bindXField = xFieldDim);
            let bindYField = yField;
            (0, vutils_1.isValid)(yFieldIndex) && (bindYField = (0, vutils_1.array)(yField)[yFieldIndex]), 
            yFieldDim && (0, vutils_1.array)(yField).includes(yFieldDim) && (bindYField = yFieldDim);
            const option = Object.assign({
                x: void 0,
                y: void 0
            }, this._getAllRelativeSeries());
            return (0, vutils_1.isString)(coordinate[bindXField]) && (0, utils_1.isAggrSpec)(coordinate[bindXField]) ? option.x = {
                field: bindXField,
                aggrType: coordinate[bindXField]
            } : option.x = (0, vutils_1.array)(bindXField).map((field => coordinate[field])), 
            (0, vutils_1.isString)(coordinate[bindYField]) && (0, utils_1.isAggrSpec)(coordinate[bindYField]) ? option.y = {
                field: bindYField,
                aggrType: coordinate[bindYField]
            } : option.y = (0, vutils_1.array)(bindYField).map((field => coordinate[field])), 
            option.getRefRelativeSeries = () => refRelativeSeries, option;
        }));
    }
    updateLayoutAttribute() {
        var _a, _b, _c;
        if (null === (_a = this._spec.visible) || void 0 === _a || _a) {
            if (!this._markerComponent) {
                const markerComponent = this._createMarkerComponent();
                markerComponent.name = null !== (_b = this._spec.name) && void 0 !== _b ? _b : this.type, 
                markerComponent.id = null !== (_c = this._spec.id) && void 0 !== _c ? _c : `${this.type}-${this.id}`, 
                this._markerComponent = markerComponent, this.getContainer().add(this._markerComponent), 
                this._markerComponent.on("*", ((event, type) => this._delegateEvent(this._markerComponent, event, type)));
            }
            this._markerLayout();
        }
        super.updateLayoutAttribute();
    }
    _getSeriesByIdOrIndex(seriesUserId, seriesIndex) {
        var _a, _b;
        let series;
        return series = null === (_a = this._option.getSeriesInUserIdOrIndex((0, vutils_1.isValid)(seriesUserId) ? [ seriesUserId ] : [], [ seriesIndex ])) || void 0 === _a ? void 0 : _a[0], 
        series || (series = null !== (_b = this._relativeSeries) && void 0 !== _b ? _b : this._getFirstSeries()), 
        series;
    }
    _bindSeries() {
        const spec = this._spec;
        this._relativeSeries = this._getSeriesByIdOrIndex(spec.relativeSeriesId, spec.relativeSeriesIndex), 
        this._startRelativeSeries = this._getSeriesByIdOrIndex(spec.startRelativeSeriesId, spec.startRelativeSeriesIndex), 
        this._endRelativeSeries = this._getSeriesByIdOrIndex(spec.endRelativeSeriesId, spec.endRelativeSeriesIndex);
    }
    initEvent() {}
    onRender(ctx) {}
    changeRegions(regions) {}
    clear() {
        super.clear(), this._firstSeries = null;
    }
    _getFirstSeries() {
        var _a;
        if (this._firstSeries) return this._firstSeries;
        const firstSeries = (0, util_1.getFirstSeries)(this._regions);
        return firstSeries ? (this._firstSeries = firstSeries, firstSeries) : (null === (_a = this._option) || void 0 === _a || _a.onError("need at least one series"), 
        null);
    }
    _getNeedClearVRenderComponents() {
        return [ this._markerComponent ];
    }
    onLayoutStart(layoutRect, chartViewRect, ctx) {
        (0, vutils_1.isNil)(this._spec.offsetX) || (this._layoutOffsetX = (0, space_1.calcLayoutNumber)(this._spec.offsetX, chartViewRect.width, chartViewRect)), 
        (0, vutils_1.isNil)(this._spec.offsetY) || (this._layoutOffsetY = (0, space_1.calcLayoutNumber)(this._spec.offsetY, chartViewRect.height, chartViewRect)), 
        super.onLayoutStart(layoutRect, chartViewRect, ctx);
    }
}

exports.BaseMarker = BaseMarker;
