// @visactor/vchart/cjs/plugin/chart/plugin-service.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.ChartPluginService = void 0;

const base_plugin_service_1 = require("../base/base-plugin-service");

class ChartPluginService extends base_plugin_service_1.BasePluginService {
    constructor(globalInstance) {
        super(), this.globalInstance = globalInstance;
    }
    onInit(chartSpec) {
        this._plugins.forEach((plugin => {
            plugin.onInit && plugin.onInit(this, chartSpec);
        }));
    }
    onBeforeResize(width, height) {
        this._plugins.forEach((plugin => {
            plugin.onBeforeResize && plugin.onBeforeResize(this, width, height);
        }));
    }
    onAfterChartSpecTransform(chartSpec, actionSource) {
        this._plugins.forEach((plugin => {
            plugin.onAfterChartSpecTransform && plugin.onAfterChartSpecTransform(this, chartSpec, actionSource);
        }));
    }
    onBeforeInitChart(chartSpec, actionSource) {
        this._plugins.forEach((plugin => {
            plugin.onBeforeInitChart && plugin.onBeforeInitChart(this, chartSpec, actionSource);
        }));
    }
}

exports.ChartPluginService = ChartPluginService;
