// @visactor/vgrammar-core/cjs/graph/mark/graphic.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerLargeSymbolsGraphic = exports.registerLargeRectsGraphic = exports.registerGlyphGraphic = exports.registerIntervalGraphic = exports.registerCellGraphic = exports.registerRichTextGraphic = exports.registerPolygonGraphic = exports.registerTextGraphic = exports.registerSymbolGraphic = exports.registerShapeGraphic = exports.registerRuleGraphic = exports.registerRect3dGraphic = exports.registerRectGraphic = exports.registerPathGraphic = exports.registerLineGraphic = exports.registerImageGraphic = exports.registerGroupGraphic = exports.registerAreaGraphic = exports.registerPyramid3dGraphic = exports.registerArc3dGraphic = exports.registerArcGraphic = exports.registerCircleGraphic = void 0;

const factory_1 = require("../../core/factory"), enums_1 = require("../enums"), large_rects_1 = require("./large-rects"), large_symbols_1 = require("./large-symbols"), vrender_kits_1 = require("@visactor/vrender-kits"), vrender_core_1 = require("@visactor/vrender-core"), registerCircleGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerCircle)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.circle, vrender_core_1.createCircle);
};

exports.registerCircleGraphic = registerCircleGraphic;

const registerArcGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerArc)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.arc, vrender_core_1.createArc);
};

exports.registerArcGraphic = registerArcGraphic;

const registerArc3dGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerArc3d)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.arc3d, vrender_core_1.createArc3d);
};

exports.registerArc3dGraphic = registerArc3dGraphic;

const registerPyramid3dGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPyramid3d)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.pyramid3d, vrender_core_1.createPyramid3d);
};

exports.registerPyramid3dGraphic = registerPyramid3dGraphic;

const registerAreaGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerArea)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.area, vrender_core_1.createArea);
};

exports.registerAreaGraphic = registerAreaGraphic;

const registerGroupGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerGroup)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.group, vrender_core_1.createGroup);
};

exports.registerGroupGraphic = registerGroupGraphic;

const registerImageGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerImage)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.image, vrender_core_1.createImage);
};

exports.registerImageGraphic = registerImageGraphic;

const registerLineGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerLine)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.line, vrender_core_1.createLine);
};

exports.registerLineGraphic = registerLineGraphic;

const registerPathGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPath)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.path, vrender_core_1.createPath);
};

exports.registerPathGraphic = registerPathGraphic;

const registerRectGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerRect)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.rect, vrender_core_1.createRect);
};

exports.registerRectGraphic = registerRectGraphic;

const registerRect3dGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerRect3d)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.rect3d, vrender_core_1.createRect3d);
};

exports.registerRect3dGraphic = registerRect3dGraphic;

const registerRuleGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerLine)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.rule, vrender_core_1.createLine);
};

exports.registerRuleGraphic = registerRuleGraphic;

const registerShapeGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPath)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.shape, vrender_core_1.createPath);
};

exports.registerShapeGraphic = registerShapeGraphic;

const registerSymbolGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerSymbol)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.symbol, vrender_core_1.createSymbol);
};

exports.registerSymbolGraphic = registerSymbolGraphic;

const registerTextGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerText)(), (0, 
    vrender_kits_1.registerRichtext)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.text, vrender_core_1.createText);
};

exports.registerTextGraphic = registerTextGraphic;

const registerPolygonGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPolygon)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.polygon, vrender_core_1.createPolygon);
};

exports.registerPolygonGraphic = registerPolygonGraphic;

const registerRichTextGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerRichtext)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.richtext, vrender_core_1.createRichText);
};

exports.registerRichTextGraphic = registerRichTextGraphic;

const registerCellGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerSymbol)(), 
    factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.cell, vrender_core_1.createSymbol);
};

exports.registerCellGraphic = registerCellGraphic;

const registerIntervalGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerRect)(), (0, 
    vrender_kits_1.registerArc)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.interval, vrender_core_1.createRect);
};

exports.registerIntervalGraphic = registerIntervalGraphic;

const registerGlyphGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerGlyph)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.glyph, vrender_core_1.createGlyph);
};

exports.registerGlyphGraphic = registerGlyphGraphic;

const registerLargeRectsGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPath)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.largeRects, (attrs => new large_rects_1.LargeRects(attrs)));
};

exports.registerLargeRectsGraphic = registerLargeRectsGraphic;

const registerLargeSymbolsGraphic = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerPath)(), factory_1.Factory.registerGraphic(enums_1.GrammarMarkType.largeSymbols, (attrs => new large_symbols_1.LargeSymbols(attrs)));
};

exports.registerLargeSymbolsGraphic = registerLargeSymbolsGraphic;
