// @visactor/vchart/esm/chart/word-cloud/3d/word-cloud-3d.js
import { SeriesTypeEnum } from "../../../series/interface/type";

import { ChartTypeEnum } from "../../interface/type";

import { BaseWordCloudChart } from "../base/base";

import { registerWordCloud3dSeries, registerWordCloudShape3dSeries } from "../../../series/word-cloud/word-cloud-3d";

import { Factory } from "../../../core/factory";

import { WordCloud3dChartSpecTransformer } from "./word-cloud-3d-transformer";

export class WordCloud3dChart extends BaseWordCloudChart {
    constructor() {
        super(...arguments), this.transformerConstructor = WordCloud3dChartSpecTransformer, 
        this.type = ChartTypeEnum.wordCloud3d, this.seriesType = SeriesTypeEnum.wordCloud3d;
    }
}

WordCloud3dChart.type = ChartTypeEnum.wordCloud3d, WordCloud3dChart.seriesType = SeriesTypeEnum.wordCloud3d, 
WordCloud3dChart.view = "singleDefault", WordCloud3dChart.transformerConstructor = WordCloud3dChartSpecTransformer;

export const registerWordCloud3dChart = () => {
    registerWordCloud3dSeries(), Factory.registerChart(WordCloud3dChart.type, WordCloud3dChart);
};

export const registerWordCloudShape3dChart = () => {
    registerWordCloudShape3dSeries(), registerWordCloud3dSeries(), Factory.registerChart(WordCloud3dChart.type, WordCloud3dChart);
};
