// @visactor/vrender-components/cjs/axis/animate/group-fade.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.GroupFadeOut = exports.GroupFadeIn = void 0;

const vrender_core_1 = require("@visactor/vrender-core");

class GroupFadeIn extends vrender_core_1.ACustomAnimate {
    constructor(from, to, duration, easing) {
        super(null, null, duration, easing);
    }
    getEndProps() {
        return {};
    }
    onBind() {
        this.target.setTheme({
            common: {
                opacity: 0
            }
        });
    }
    onEnd() {
        this.target.setTheme({
            common: {
                opacity: 1
            }
        });
    }
    onUpdate(end, ratio, out) {
        this.target.setTheme({
            common: {
                opacity: ratio
            }
        });
    }
}

exports.GroupFadeIn = GroupFadeIn;

class GroupFadeOut extends vrender_core_1.ACustomAnimate {
    constructor(from, to, duration, easing) {
        super(null, null, duration, easing);
    }
    getEndProps() {
        return {};
    }
    onBind() {
        this.target.setTheme({
            common: {
                opacity: 1
            }
        });
    }
    onEnd() {
        this.target.setTheme({
            common: {
                opacity: 0
            }
        });
    }
    onUpdate(end, ratio, out) {
        this.target.setTheme({
            common: {
                opacity: 1 - ratio
            }
        });
    }
}

exports.GroupFadeOut = GroupFadeOut;
