// @visactor/vchart/cjs/chart/polar/polar-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.PolarChartSpecTransformer = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("../base"), constant_1 = require("../../component/data-zoom/constant");

class PolarChartSpecTransformer extends base_1.BaseChartSpecTransformer {
    _isValidSeries(type) {
        return !this.seriesType || type === this.seriesType;
    }
    getIndicatorSpec(spec) {
        var _a, _b, _c;
        const indicatorSpec = (0, vutils_1.array)(spec.indicator), limitRatio = null !== (_a = spec.innerRadius) && void 0 !== _a ? _a : null === (_c = null === (_b = spec.series) || void 0 === _b ? void 0 : _b[0]) || void 0 === _c ? void 0 : _c.innerRadius;
        return (0, vutils_1.isValid)(limitRatio) && indicatorSpec.forEach((indicator => {
            (0, vutils_1.isNil)(indicator.limitRatio) && (indicator.limitRatio = limitRatio);
        })), indicatorSpec;
    }
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            radius: spec.radius,
            outerRadius: spec.outerRadius,
            innerRadius: spec.innerRadius,
            startAngle: spec.startAngle,
            endAngle: spec.endAngle,
            sortDataByAxis: spec.sortDataByAxis
        }), seriesType = this.seriesType;
        return seriesType && (series.type = seriesType, series[seriesType] = spec[seriesType]), 
        series;
    }
    transformSpec(spec) {
        super.transformSpec(spec), (0, vutils_1.isArray)(spec.dataZoom) && spec.dataZoom.length > 0 && spec.dataZoom.forEach((zoom => {
            zoom.filterMode === constant_1.IFilterMode.axis && (zoom.filterMode = constant_1.IFilterMode.filter);
        }));
        const defaultSeriesSpec = this._getDefaultSeriesSpec(spec);
        spec.series && 0 !== spec.series.length ? spec.series.forEach((s => {
            this._isValidSeries(s.type) && Object.keys(defaultSeriesSpec).forEach((k => {
                k in s || (s[k] = defaultSeriesSpec[k]);
            }));
        })) : spec.series = [ defaultSeriesSpec ], (0, vutils_1.isValid)(spec.indicator) && (spec.indicator = this.getIndicatorSpec(spec));
    }
}

exports.PolarChartSpecTransformer = PolarChartSpecTransformer;
