// @visactor/vgrammar-util/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.parseViewBox = exports.regressionLinear = exports.toPercent = exports.splitAccessPath = exports.isEqual = exports.field = exports.extent = exports.error = exports.ascending = exports.compare = exports.emptyObject = exports.falsy = exports.truthy = exports.one = exports.zero = exports.identity = exports.id = exports.getter = exports.accessorFields = exports.accessorName = exports.accessor = void 0;

var accessor_1 = require("./accessor");

Object.defineProperty(exports, "accessor", {
    enumerable: !0,
    get: function() {
        return accessor_1.accessor;
    }
}), Object.defineProperty(exports, "accessorName", {
    enumerable: !0,
    get: function() {
        return accessor_1.accessorName;
    }
}), Object.defineProperty(exports, "accessorFields", {
    enumerable: !0,
    get: function() {
        return accessor_1.accessorFields;
    }
});

var getter_1 = require("./getter");

Object.defineProperty(exports, "getter", {
    enumerable: !0,
    get: function() {
        return getter_1.getter;
    }
});

var accessors_1 = require("./accessors");

Object.defineProperty(exports, "id", {
    enumerable: !0,
    get: function() {
        return accessors_1.id;
    }
}), Object.defineProperty(exports, "identity", {
    enumerable: !0,
    get: function() {
        return accessors_1.identity;
    }
}), Object.defineProperty(exports, "zero", {
    enumerable: !0,
    get: function() {
        return accessors_1.zero;
    }
}), Object.defineProperty(exports, "one", {
    enumerable: !0,
    get: function() {
        return accessors_1.one;
    }
}), Object.defineProperty(exports, "truthy", {
    enumerable: !0,
    get: function() {
        return accessors_1.truthy;
    }
}), Object.defineProperty(exports, "falsy", {
    enumerable: !0,
    get: function() {
        return accessors_1.falsy;
    }
}), Object.defineProperty(exports, "emptyObject", {
    enumerable: !0,
    get: function() {
        return accessors_1.emptyObject;
    }
});

var compare_1 = require("./compare");

Object.defineProperty(exports, "compare", {
    enumerable: !0,
    get: function() {
        return compare_1.compare;
    }
}), Object.defineProperty(exports, "ascending", {
    enumerable: !0,
    get: function() {
        return compare_1.ascending;
    }
});

var error_1 = require("./error");

Object.defineProperty(exports, "error", {
    enumerable: !0,
    get: function() {
        return error_1.error;
    }
});

var extent_1 = require("./extent");

Object.defineProperty(exports, "extent", {
    enumerable: !0,
    get: function() {
        return extent_1.extent;
    }
});

var field_1 = require("./field");

Object.defineProperty(exports, "field", {
    enumerable: !0,
    get: function() {
        return field_1.field;
    }
});

var isEqual_1 = require("./isEqual");

Object.defineProperty(exports, "isEqual", {
    enumerable: !0,
    get: function() {
        return isEqual_1.isEqual;
    }
});

var splitAccessPath_1 = require("./splitAccessPath");

Object.defineProperty(exports, "splitAccessPath", {
    enumerable: !0,
    get: function() {
        return splitAccessPath_1.splitAccessPath;
    }
});

var toPercent_1 = require("./toPercent");

Object.defineProperty(exports, "toPercent", {
    enumerable: !0,
    get: function() {
        return toPercent_1.toPercent;
    }
}), __exportStar(require("./types"), exports);

var regression_linear_1 = require("./regression-linear");

Object.defineProperty(exports, "regressionLinear", {
    enumerable: !0,
    get: function() {
        return regression_linear_1.regressionLinear;
    }
});

var view_box_1 = require("./view-box");

Object.defineProperty(exports, "parseViewBox", {
    enumerable: !0,
    get: function() {
        return view_box_1.parseViewBox;
    }
}), __exportStar(require("./direction"), exports);
