// @visactor/vchart/cjs/component/common/trigger/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.getDefaultTriggerEventByMode = void 0;

const common_1 = require("../../../typings/spec/common"), env_1 = require("../../../util/env"), desktop_1 = require("./desktop"), mobile_1 = require("./mobile");

function getDefaultTriggerEventByMode(mode) {
    return mode === common_1.RenderModeEnum["desktop-browser"] || mode === common_1.RenderModeEnum["desktop-miniApp"] ? {
        start: "pointerdown",
        move: "pointermove",
        end: "pointerup",
        zoom: "wheel",
        zoomEnd: "pointerup",
        scroll: "wheel",
        trigger: desktop_1.DeskTopTrigger
    } : (0, env_1.isMobileLikeMode)(mode) || (0, env_1.isMiniAppLikeMode)(mode) ? {
        start: "pointerdown",
        move: "pointermove",
        end: "pointerup",
        zoom: "pinch",
        zoomEnd: "pinchend",
        scroll: "pan",
        scrollEnd: "panend",
        trigger: mobile_1.MobileTrigger
    } : null;
}

exports.getDefaultTriggerEventByMode = getDefaultTriggerEventByMode;
