// @visactor/vchart/esm/component/marker/mark-point/mark-point.js
var __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

import { DataView } from "@visactor/vdataset";

import { ComponentTypeEnum } from "../../interface/type";

import { markerAggregation } from "../../../data/transforms/aggregation";

import { computeClipRange, coordinateLayout, positionLayout, transformLabelAttributes, xyLayout } from "../utils";

import { registerDataSetInstanceTransform } from "../../../data/register";

import { MarkPoint as MarkPointComponent } from "@visactor/vrender-components";

import { isEmpty, isValid, isArray } from "@visactor/vutils";

import { transformToGraphic } from "../../../util/style";

import { BaseMarker } from "../base-marker";

import { LayoutZIndex } from "../../../constant";

import { Factory } from "../../../core/factory";

export class MarkPoint extends BaseMarker {
    constructor() {
        super(...arguments), this.type = ComponentTypeEnum.markPoint, this.name = ComponentTypeEnum.markPoint, 
        this.specKey = "markPoint", this.layoutZIndex = LayoutZIndex.MarkPoint;
    }
    static getSpecInfo(chartSpec) {
        const markPointSpec = chartSpec[this.specKey];
        if (isEmpty(markPointSpec)) return;
        if (!isArray(markPointSpec) && !1 !== markPointSpec.visible) return [ {
            spec: markPointSpec,
            specPath: [ this.specKey ],
            type: ComponentTypeEnum.markPoint
        } ];
        const specInfos = [];
        return markPointSpec.forEach(((m, i) => {
            !1 !== m.visible && specInfos.push({
                spec: m,
                specIndex: i,
                specPath: [ this.specKey, i ],
                type: ComponentTypeEnum.markPoint
            });
        })), specInfos;
    }
    _createMarkerComponent() {
        var _a, _b;
        const {itemContent: itemContent = {}, itemLine: itemLine = {}} = this._spec, {text: label = {}, symbol: symbol, image: image, richText: richText} = itemContent, restItemContent = __rest(itemContent, [ "text", "symbol", "image", "richText" ]), markPointAttrs = {
            zIndex: this.layoutZIndex,
            interactive: null !== (_a = this._spec.interactive) && void 0 !== _a && _a,
            position: {
                x: 0,
                y: 0
            },
            clipInRange: null !== (_b = this._spec.clip) && void 0 !== _b && _b,
            itemContent: Object.assign({
                symbolStyle: transformToGraphic(null == symbol ? void 0 : symbol.style),
                imageStyle: null == image ? void 0 : image.style,
                textStyle: transformLabelAttributes(label),
                richTextStyle: null == richText ? void 0 : richText.style
            }, restItemContent)
        }, {visible: visible, line: line = {}} = itemLine, restItemLine = __rest(itemLine, [ "visible", "line" ]);
        markPointAttrs.itemLine = !1 !== visible ? Object.assign(Object.assign({}, restItemLine), {
            visible: !0,
            lineStyle: transformToGraphic(line.style)
        }) : {
            visible: !1
        };
        return new MarkPointComponent(markPointAttrs);
    }
    _markerLayout() {
        var _a, _b, _c, _d, _e, _f;
        const spec = this._spec, data = this._markerData, relativeSeries = this._relativeSeries, isXYLayout = "x" in spec && "y" in spec, isCoordinateLayout = "coordinate" in spec, isPositionLayout = "position" in spec, autoRange = null !== (_a = null == spec ? void 0 : spec.autoRange) && void 0 !== _a && _a;
        let point;
        isXYLayout ? point = xyLayout(data, relativeSeries, relativeSeries, relativeSeries, autoRange)[0][0] : isCoordinateLayout ? point = coordinateLayout(data, relativeSeries, autoRange, spec.coordinatesOffset)[0] : isPositionLayout && (point = positionLayout([ spec.position ], relativeSeries, spec.regionRelative)[0]);
        const seriesData = this._relativeSeries.getViewData().latestData, dataPoints = data ? data.latestData[0].latestData ? data.latestData[0].latestData : data.latestData : seriesData;
        let limitRect;
        if (spec.clip || (null === (_b = spec.itemContent) || void 0 === _b ? void 0 : _b.confine)) {
            const {minX: minX, maxX: maxX, minY: minY, maxY: maxY} = computeClipRange([ relativeSeries.getRegion() ]);
            limitRect = {
                x: minX,
                y: minY,
                width: maxX - minX,
                height: maxY - minY
            };
        }
        if (this._markerComponent) {
            const attribute = null !== (_c = this._markerComponent.attribute) && void 0 !== _c ? _c : {}, textStyle = null !== (_e = null === (_d = attribute.itemContent) || void 0 === _d ? void 0 : _d.textStyle) && void 0 !== _e ? _e : {};
            this._markerComponent.setAttributes({
                position: point,
                itemContent: Object.assign(Object.assign({}, attribute.itemContent), {
                    textStyle: Object.assign(Object.assign({}, textStyle), {
                        text: (null === (_f = this._spec.itemContent.text) || void 0 === _f ? void 0 : _f.formatMethod) ? this._spec.itemContent.text.formatMethod(dataPoints, seriesData) : textStyle.text
                    })
                }),
                limitRect: limitRect,
                dx: this._layoutOffsetX,
                dy: this._layoutOffsetY
            });
        }
    }
    _initDataView() {
        const spec = this._spec, relativeSeries = this._relativeSeries, isXYProcess = isValid(spec.x) && isValid(spec.y), isCoordinateProcess = isValid(spec.coordinate);
        if (!isCoordinateProcess && !isXYProcess) return;
        let options;
        registerDataSetInstanceTransform(this._option.dataSet, "markerAggregation", markerAggregation), 
        isXYProcess ? options = [ this._processSpecXY(spec.x, spec.y) ] : isCoordinateProcess && (options = this._processSpecCoo(spec));
        const data = new DataView(this._option.dataSet, {
            name: `${this.type}_${this.id}_data`
        });
        data.parse([ relativeSeries.getViewData() ], {
            type: "dataview"
        }), data.transform({
            type: "markerAggregation",
            options: options
        }), data.target.on("change", (() => {
            this._markerLayout();
        })), this._markerData = data;
    }
}

MarkPoint.type = ComponentTypeEnum.markPoint, MarkPoint.specKey = "markPoint";

export const registerMarkPoint = () => {
    Factory.registerComponent(MarkPoint.type, MarkPoint);
};
