// @visactor/vchart/cjs/env/env.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerWXEnv = exports.registerTaroEnv = exports.registerNodeEnv = exports.registerLynxEnv = exports.registerLarkEnv = exports.registerBrowserEnv = exports.registerAllEnv = void 0;

const vrender_kits_1 = require("@visactor/vrender-kits"), vrender_core_1 = require("@visactor/vrender-core"), registerAllEnv = () => {
    (0, vrender_kits_1.initAllEnv)();
};

exports.registerAllEnv = registerAllEnv;

const registerBrowserEnv = () => {
    (0, vrender_kits_1.loadBrowserEnv)(vrender_core_1.container);
};

exports.registerBrowserEnv = registerBrowserEnv;

const registerLarkEnv = () => {
    (0, vrender_kits_1.loadFeishuEnv)(vrender_core_1.container);
};

exports.registerLarkEnv = registerLarkEnv;

const registerLynxEnv = () => {
    (0, vrender_kits_1.loadLynxEnv)(vrender_core_1.container);
};

exports.registerLynxEnv = registerLynxEnv;

const registerNodeEnv = () => {
    (0, vrender_kits_1.loadNodeEnv)(vrender_core_1.container);
};

exports.registerNodeEnv = registerNodeEnv;

const registerTaroEnv = () => {
    (0, vrender_kits_1.loadTaroEnv)(vrender_core_1.container);
};

exports.registerTaroEnv = registerTaroEnv;

const registerWXEnv = () => {
    (0, vrender_kits_1.loadWxEnv)(vrender_core_1.container);
};

exports.registerWXEnv = registerWXEnv;
