// @visactor/vrender-core/cjs/graphic/arc3d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.createArc3d = exports.Arc3d = void 0;

const arc_1 = require("./arc"), theme_1 = require("./theme"), application_1 = require("../application"), utils_1 = require("../common/utils"), constants_1 = require("./constants"), graphic_1 = require("./graphic");

class Arc3d extends arc_1.Arc {
    constructor(params) {
        super(params), this.type = "arc3d", this.numberType = constants_1.ARC3D_NUMBER_TYPE;
    }
    doUpdateAABBBounds() {
        const polygonTheme = (0, theme_1.getTheme)(this).arc;
        this._AABBBounds.setValue(1 / 0, 1 / 0, -1 / 0, -1 / 0);
        const attribute = this.attribute, bounds = application_1.application.graphicService.updateArc3dAABBBounds(attribute, (0, 
        theme_1.getTheme)(this).polygon, this._AABBBounds, this), {boundsPadding: boundsPadding = polygonTheme.boundsPadding} = attribute, paddingArray = (0, 
        utils_1.parsePadding)(boundsPadding);
        return paddingArray && bounds.expand(paddingArray), this.clearUpdateBoundTag(), 
        this._AABBBounds;
    }
    getNoWorkAnimateAttr() {
        return Arc3d.NOWORK_ANIMATE_ATTR;
    }
}

function createArc3d(attributes) {
    return new Arc3d(attributes);
}

exports.Arc3d = Arc3d, Arc3d.NOWORK_ANIMATE_ATTR = Object.assign({
    cap: 1
}, graphic_1.NOWORK_ANIMATE_ATTR), exports.createArc3d = createArc3d;
