// @visactor/vrender-core/cjs/graphic/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.SVG_PARSE_ATTRIBUTE_MAP_KEYS = exports.SVG_PARSE_ATTRIBUTE_MAP = exports.SVG_ATTRIBUTE_MAP_KEYS = exports.SVG_ATTRIBUTE_MAP = exports.GraphicCreator = exports.GraphicService = exports.TEXT_NUMBER_TYPE = exports.SYMBOL_NUMBER_TYPE = exports.RICHTEXT_NUMBER_TYPE = exports.RECT3D_NUMBER_TYPE = exports.RECT_NUMBER_TYPE = exports.PYRAMID3D_NUMBER_TYPE = exports.POLYGON_NUMBER_TYPE = exports.PATH_NUMBER_TYPE = exports.LINE_NUMBER_TYPE = exports.IMAGE_NUMBER_TYPE = exports.GROUP_NUMBER_TYPE = exports.GLYPH_NUMBER_TYPE = exports.CIRCLE_NUMBER_TYPE = exports.AREA_NUMBER_TYPE = exports.ARC3D_NUMBER_TYPE = exports.ARC_NUMBER_TYPE = void 0;

const tools_1 = require("./tools");

exports.ARC_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.ARC3D_NUMBER_TYPE = (0, 
tools_1.genNumberType)(), exports.AREA_NUMBER_TYPE = (0, tools_1.genNumberType)(), 
exports.CIRCLE_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.GLYPH_NUMBER_TYPE = (0, 
tools_1.genNumberType)(), exports.GROUP_NUMBER_TYPE = (0, tools_1.genNumberType)(), 
exports.IMAGE_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.LINE_NUMBER_TYPE = (0, 
tools_1.genNumberType)(), exports.PATH_NUMBER_TYPE = (0, tools_1.genNumberType)(), 
exports.POLYGON_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.PYRAMID3D_NUMBER_TYPE = (0, 
tools_1.genNumberType)(), exports.RECT_NUMBER_TYPE = (0, tools_1.genNumberType)(), 
exports.RECT3D_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.RICHTEXT_NUMBER_TYPE = (0, 
tools_1.genNumberType)(), exports.SYMBOL_NUMBER_TYPE = (0, tools_1.genNumberType)(), 
exports.TEXT_NUMBER_TYPE = (0, tools_1.genNumberType)(), exports.GraphicService = Symbol.for("GraphicService"), 
exports.GraphicCreator = Symbol.for("GraphicCreator"), exports.SVG_ATTRIBUTE_MAP = {
    "stroke-linecap": "lineCap",
    "stroke-linejoin": "lineJoin",
    "stroke-dasharray": "lineDash",
    "stroke-dashoffset": "lineDashOffset",
    "stroke-width": "lineWidth",
    "fill-opacity": "fillOpacity",
    "stroke-opacity": "strokeOpacity"
}, exports.SVG_ATTRIBUTE_MAP_KEYS = Object.keys(exports.SVG_ATTRIBUTE_MAP), exports.SVG_PARSE_ATTRIBUTE_MAP = {
    "stroke-linecap": "lineCap",
    "stroke-linejoin": "lineJoin",
    "stroke-dasharray": "lineDash",
    "stroke-dashoffset": "lineDashOffset",
    "stroke-width": "lineWidth",
    "fill-opacity": "fillOpacity",
    "stroke-opacity": "strokeOpacity",
    stroke: "stroke",
    fill: "fill"
}, exports.SVG_PARSE_ATTRIBUTE_MAP_KEYS = Object.keys(exports.SVG_PARSE_ATTRIBUTE_MAP);
