// @visactor/vrender-core/cjs/graphic/path.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.createPath = exports.Path = void 0;

const vutils_1 = require("@visactor/vutils"), graphic_1 = require("./graphic"), utils_1 = require("../common/utils"), custom_path2d_1 = require("../common/custom-path2d"), theme_1 = require("./theme"), application_1 = require("../application"), constants_1 = require("./constants"), PATH_UPDATE_TAG_KEY = [ "path", "customPath", ...graphic_1.GRAPHIC_UPDATE_TAG_KEY ];

class Path extends graphic_1.Graphic {
    constructor(params) {
        super(params), this.type = "path", this.numberType = constants_1.PATH_NUMBER_TYPE;
    }
    get pathShape() {
        return this.tryUpdateAABBBounds(), this.getParsedPathShape();
    }
    isValid() {
        return super.isValid() && this._isValid();
    }
    _isValid() {
        const {path: path} = this.attribute;
        return null != path && "" !== path;
    }
    getParsedPathShape() {
        const pathTheme = (0, theme_1.getTheme)(this).path;
        if (!this.valid) return pathTheme.path;
        const attribute = this.attribute;
        return attribute.path instanceof custom_path2d_1.CustomPath2D ? attribute.path : ((0, 
        vutils_1.isNil)(this.cache) && this.doUpdatePathShape(), this.cache instanceof custom_path2d_1.CustomPath2D ? this.cache : pathTheme.path);
    }
    doUpdateAABBBounds() {
        const pathTheme = (0, theme_1.getTheme)(this).path;
        this.doUpdatePathShape(), this._AABBBounds.setValue(1 / 0, 1 / 0, -1 / 0, -1 / 0);
        const attribute = this.attribute, bounds = application_1.application.graphicService.updatePathAABBBounds(attribute, (0, 
        theme_1.getTheme)(this).path, this._AABBBounds, this), {boundsPadding: boundsPadding = pathTheme.boundsPadding} = attribute, paddingArray = (0, 
        utils_1.parsePadding)(boundsPadding);
        return paddingArray && bounds.expand(paddingArray), this.clearUpdateBoundTag(), 
        bounds;
    }
    doUpdatePathShape() {
        const attribute = this.attribute;
        (0, vutils_1.isString)(attribute.path, !0) ? this.cache = (new custom_path2d_1.CustomPath2D).fromString(attribute.path) : attribute.customPath && (this.cache = new custom_path2d_1.CustomPath2D, 
        attribute.customPath(this.cache, this));
    }
    tryUpdateOBBBounds() {
        throw new Error("暂不支持");
    }
    getDefaultAttribute(name) {
        return (0, theme_1.getTheme)(this).path[name];
    }
    needUpdateTags(keys) {
        return super.needUpdateTags(keys, PATH_UPDATE_TAG_KEY);
    }
    needUpdateTag(key) {
        return super.needUpdateTag(key, PATH_UPDATE_TAG_KEY);
    }
    toCustomPath() {
        return (new custom_path2d_1.CustomPath2D).fromCustomPath2D(this.getParsedPathShape(), 0, 0);
    }
    clone() {
        return new Path(Object.assign({}, this.attribute));
    }
    getNoWorkAnimateAttr() {
        return Path.NOWORK_ANIMATE_ATTR;
    }
}

function createPath(attributes) {
    return new Path(attributes);
}

exports.Path = Path, Path.NOWORK_ANIMATE_ATTR = Object.assign({
    path: 1,
    customPath: 1
}, graphic_1.NOWORK_ANIMATE_ATTR), exports.createPath = createPath;
