// @visactor/vchart/esm/chart/correlation/correlation-transformer.js
import { SeriesTypeEnum } from "../../series";

import { BaseChartSpecTransformer } from "../base";

export class CorrelationChartSpecTransformer extends BaseChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            categoryField: spec.categoryField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            sizeField: spec.sizeField,
            sizeRange: spec.sizeRange,
            centerX: spec.centerX,
            centerY: spec.centerY,
            innerRadius: spec.innerRadius,
            outerRadius: spec.outerRadius,
            startAngle: spec.startAngle,
            endAngle: spec.endAngle,
            ripplePoint: spec.ripplePoint,
            centerPoint: spec.centerPoint,
            centerLabel: spec.centerLabel,
            nodePoint: spec.nodePoint,
            label: spec.label
        }), seriesType = SeriesTypeEnum.correlation;
        return series.type = seriesType, series[seriesType] = spec[seriesType], series;
    }
    transformSpec(spec) {
        super.transformSpec(spec);
        const defaultSeriesSpec = this._getDefaultSeriesSpec(spec);
        spec.series && 0 !== spec.series.length ? spec.series.forEach((s => {
            this._isValidSeries(s.type) && Object.keys(defaultSeriesSpec).forEach((k => {
                k in s || (s[k] = defaultSeriesSpec[k]);
            }));
        })) : spec.series = [ defaultSeriesSpec ];
    }
}
