// @visactor/vchart/cjs/component/data-zoom/data-zoom/data-zoom-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.DataZoomSpecTransformer = void 0;

const util_1 = require("../../../util"), base_1 = require("../../base");

class DataZoomSpecTransformer extends base_1.BaseComponentSpecTransformer {
    _mergeThemeToSpec(spec, chartSpec) {
        const theme = this._theme;
        let newSpec = spec;
        if (this._shouldMergeThemeToSpec()) {
            const merge = originalSpec => {
                const result = (0, util_1.mergeSpec)({
                    selectedBackgroundChart: {
                        line: {},
                        area: {}
                    }
                }, this._theme, originalSpec), {selectedBackgroundChart: selectedBackgroundChart = {}} = originalSpec, {line: line, area: area} = selectedBackgroundChart;
                return line && !1 !== line.visible && (result.selectedBackgroundChart.line.style = Object.assign(Object.assign({}, result.selectedBackgroundChart.line.style), {
                    visible: !0
                })), area && !1 !== area.visible && (result.selectedBackgroundChart.area.style = Object.assign(Object.assign({}, result.selectedBackgroundChart.area.style), {
                    visible: !0
                })), result;
            }, baseSpec = spec;
            newSpec = (0, util_1.isArray)(baseSpec) ? baseSpec.map((spec => merge(spec))) : merge(baseSpec);
        }
        return this._adjustPadding(newSpec), {
            spec: newSpec,
            theme: theme
        };
    }
}

exports.DataZoomSpecTransformer = DataZoomSpecTransformer;
