// @visactor/vgrammar-core/es/interactions/element-highlight-by-legend.js
import { InteractionStateEnum } from "../graph/enums";

import { BaseInteraction } from "./base";

import { LegendEvent } from "@visactor/vrender-components";

import { isNil } from "@visactor/vutils";

import { generateFilterValue } from "./utils";

export class ElementHighlightByLegend extends BaseInteraction {
    constructor(view, options) {
        super(view, options), this.type = ElementHighlightByLegend.type, this.handleStart = (e, element) => {
            var _a, _b;
            const itemKey = null === (_b = null === (_a = e.detail) || void 0 === _a ? void 0 : _a.data) || void 0 === _b ? void 0 : _b.id;
            if (isNil(itemKey)) return;
            const filterValue = generateFilterValue(this.options);
            this._marks.forEach((mark => {
                mark.elements.forEach((el => {
                    filterValue(el) === itemKey ? (el.removeState(this.options.blurState), el.addState(this.options.highlightState)) : (el.removeState(this.options.highlightState), 
                    el.addState(this.options.blurState));
                }));
            }));
        }, this.handleReset = e => {
            this._marks.forEach((mark => {
                mark.elements.forEach((el => {
                    el.removeState(this.options.blurState), el.removeState(this.options.highlightState);
                }));
            }));
        }, this.options = Object.assign({}, ElementHighlightByLegend.defaultOptions, options), 
        this._marks = view.getMarksBySelector(this.options.selector);
    }
    getEvents() {
        return [ {
            type: LegendEvent.legendItemHover,
            handler: this.handleStart
        }, {
            type: LegendEvent.legendItemUnHover,
            handler: this.handleReset
        } ];
    }
}

ElementHighlightByLegend.type = "element-highlight-by-legend", ElementHighlightByLegend.defaultOptions = {
    highlightState: InteractionStateEnum.highlight,
    blurState: InteractionStateEnum.blur,
    filterType: "groupKey"
};
