// @visactor/vgrammar-core/es/interactions/element-highlight-by-group.js
import { isNil } from "@visactor/vutils";

import { BaseInteraction } from "./base";

import { InteractionStateEnum } from "../graph/enums";

export class ElementHighlightByGroup extends BaseInteraction {
    constructor(view, options) {
        super(view, options), this.type = ElementHighlightByGroup.type, this.handleStart = e => {
            if (e.element && this._marks && this._marks.includes(e.element.mark)) {
                const highlightKey = e.element.groupKey;
                if (isNil(highlightKey)) return;
                this._marks.forEach((mark => {
                    mark.elements.forEach((el => {
                        el.groupKey === highlightKey ? (el.removeState(this.options.blurState), el.addState(this.options.highlightState)) : (el.removeState(this.options.highlightState), 
                        el.addState(this.options.blurState));
                    }));
                }));
            }
        }, this.handleReset = e => {
            e.element && this._marks && this._marks.includes(e.element.mark) && this.clearPrevElements();
        }, this.options = Object.assign({}, ElementHighlightByGroup.defaultOptions, options), 
        this._marks = view.getMarksBySelector(this.options.selector);
    }
    getEvents() {
        return [ {
            type: this.options.trigger,
            handler: this.handleStart
        }, {
            type: this.options.resetTrigger,
            handler: this.handleReset
        } ];
    }
    clearPrevElements() {
        this._marks.forEach((mark => {
            mark.elements.forEach((el => {
                el.removeState(this.options.highlightState), el.removeState(this.options.blurState);
            }));
        }));
    }
}

ElementHighlightByGroup.type = "element-highlight-by-group", ElementHighlightByGroup.defaultOptions = {
    highlightState: InteractionStateEnum.highlight,
    blurState: InteractionStateEnum.blur,
    trigger: "pointerover",
    resetTrigger: "pointerout"
};
