// @visactor/vgrammar-core/es/interactions/element-highlight-by-name.js
import { InteractionStateEnum } from "../graph/enums";

import { BaseInteraction } from "./base";

import { array, isNil } from "@visactor/vutils";

import { generateFilterValue } from "./utils";

export class ElementHighlightByName extends BaseInteraction {
    constructor(view, options) {
        super(view, options), this.type = ElementHighlightByName.type, this.handleStart = (e, element) => {
            if (this.options.shouldStart ? this.options.shouldStart(e) : this._filterByName(e)) {
                const itemKey = this._parseTargetKey(e, element);
                if (isNil(itemKey)) return;
                const filterValue = generateFilterValue(this.options);
                this._marks.forEach((mark => {
                    mark.elements.forEach((el => {
                        filterValue(el) === itemKey ? (el.removeState(this.options.blurState), el.addState(this.options.highlightState)) : (el.removeState(this.options.highlightState), 
                        el.addState(this.options.blurState));
                    }));
                }));
            }
        }, this.handleReset = e => {
            (this.options.shouldReset ? this.options.shouldReset(e) : this._filterByName(e)) && this._marks.forEach((mark => {
                mark.elements.forEach((el => {
                    el.removeState(this.options.blurState), el.removeState(this.options.highlightState);
                }));
            }));
        }, this.options = Object.assign({}, ElementHighlightByName.defaultOptions, options), 
        this._marks = view.getMarksBySelector(this.options.selector);
    }
    getEvents() {
        return [ {
            type: this.options.trigger,
            handler: this.handleStart
        }, {
            type: this.options.resetTrigger,
            handler: this.handleReset
        } ];
    }
    _filterByName(e) {
        var _a;
        const names = array(this.options.graphicName);
        return (null === (_a = null == e ? void 0 : e.target) || void 0 === _a ? void 0 : _a.name) && names.includes(e.target.name);
    }
    _parseTargetKey(e, element) {
        return this.options.parseData ? this.options.parseData(e) : "text" === e.target.type ? e.target.attribute.text : null;
    }
}

ElementHighlightByName.type = "element-highlight-by-name", ElementHighlightByName.defaultOptions = {
    trigger: "pointerover",
    resetTrigger: "pointerout",
    highlightState: InteractionStateEnum.highlight,
    blurState: InteractionStateEnum.blur,
    filterType: "groupKey"
};
