// @visactor/vchart/cjs/series/word-cloud/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.SHAPE_TYPE = exports.DEFAULT_TEXT_BASE_LINE = exports.DEFAULT_TEXT_ALIGN = exports.DEFAULT_FONT_PADDING = exports.DEFAULT_RANDOM = exports.DEFAULT_ZOOM_TO_FIT = exports.DEFAULT_MASK_SHAPE = exports.DEFAULT_FONT_WEIGHT_RANGE = exports.DEFAULT_FONTSIZE_RANGE = exports.DEFAULT_MIN_FONT_SIZE = exports.DEFAULT_DRAW_OUT_OF_BOUND = exports.DEFAULT_ROTATE_ANGLES = void 0, 
exports.DEFAULT_ROTATE_ANGLES = [ 0 ], exports.DEFAULT_DRAW_OUT_OF_BOUND = "hidden", 
exports.DEFAULT_MIN_FONT_SIZE = 20, exports.DEFAULT_FONTSIZE_RANGE = [ 20, 40 ], 
exports.DEFAULT_FONT_WEIGHT_RANGE = [ 200, 500 ], exports.DEFAULT_MASK_SHAPE = "circle", 
exports.DEFAULT_ZOOM_TO_FIT = {
    shrink: !1,
    enlarge: !1,
    fontSizeLimitMin: 0
}, exports.DEFAULT_RANDOM = !0, exports.DEFAULT_FONT_PADDING = 1, exports.DEFAULT_TEXT_ALIGN = "center", 
exports.DEFAULT_TEXT_BASE_LINE = "alphabetic", exports.SHAPE_TYPE = [ "triangleForward", "triangle", "diamond", "square", "star", "cardioid", "circle", "pentagon" ];
