// @visactor/vchart/esm/series/index.js
import { WaterfallSeries, registerWaterfallSeries } from "./waterfall/waterfall";

import { BoxPlotSeries, registerBoxplotSeries } from "./box-plot/box-plot";

import { LineSeries, registerLineSeries } from "./line/line";

import { BarSeries, registerBarSeries } from "./bar/bar";

import { Bar3dSeries, registerBar3dSeries } from "./bar/bar-3d";

import { RangeColumnSeries, registerRangeColumnSeries } from "./range-column/range-column";

import { RangeColumn3dSeries, registerRangeColumn3dSeries } from "./range-column/3d/range-column-3d";

import { RangeAreaSeries, registerRangeAreaSeries } from "./range-area/range-area";

import { MapSeries, registerMapSeries } from "./map/map";

import { PieSeries, registerPieSeries } from "./pie/pie";

import { Pie3dSeries, registerPie3dSeries } from "./pie/3d/pie-3d";

import { ScatterSeries, registerScatterSeries } from "./scatter/scatter";

import { RoseSeries, registerRoseSeries } from "./rose/rose";

import { RadarSeries, registerRadarSeries } from "./radar/radar";

import { AreaSeries, registerAreaSeries } from "./area/area";

import { DotSeries, registerDotSeries } from "./dot/dot";

import { LinkSeries, registerLinkSeries } from "./link/link";

import { CircularProgressSeries, registerCircularProgressSeries } from "./progress/circular/circular";

import { LinearProgressSeries, registerLinearProgressSeries } from "./progress/linear/linear";

import { WordCloudSeries, registerWordCloudSeries } from "./word-cloud/word-cloud";

import { WordCloud3dSeries, registerWordCloud3dSeries } from "./word-cloud/word-cloud-3d";

import { FunnelSeries, registerFunnelSeries } from "./funnel/funnel";

import { Funnel3dSeries, registerFunnel3dSeries } from "./funnel/3d/funnel-3d";

import { SunburstSeries, registerSunBurstSeries } from "./sunburst/sunburst";

import { CirclePackingSeries, registerCirclePackingSeries } from "./circle-packing/circle-packing";

import { SankeySeries, registerSankeySeries } from "./sankey/sankey";

import { TreemapSeries, registerTreemapSeries } from "./treemap/treemap";

import { GaugePointerSeries, GaugeSeries, registerGaugePointerSeries, registerGaugeSeries } from "./gauge";

import { HeatmapSeries, registerHeatmapSeries } from "./heatmap/heatmap";

import { CorrelationSeries } from "./correlation/correlation";

import { BaseSeries } from "./base/base-series";

import { CartesianSeries } from "./cartesian";

import { PolarSeries } from "./polar/polar";

import { ProgressLikeSeries } from "./polar/progress-like";

import { RoseLikeSeries } from "./polar/rose-like";

export { WaterfallSeries, Bar3dSeries, BarSeries, BoxPlotSeries, LineSeries, RadarSeries, RangeAreaSeries, RangeColumn3dSeries, RangeColumnSeries, MapSeries, Pie3dSeries, PieSeries, SankeySeries, ScatterSeries, SunburstSeries, RoseLikeSeries, RoseSeries, AreaSeries, DotSeries, LinearProgressSeries, LinkSeries, CirclePackingSeries, CircularProgressSeries, WordCloud3dSeries, WordCloudSeries, Funnel3dSeries, FunnelSeries, TreemapSeries, GaugePointerSeries, GaugeSeries, HeatmapSeries, BaseSeries, CartesianSeries, PolarSeries, ProgressLikeSeries, CorrelationSeries };

export { registerAreaSeries, registerBar3dSeries, registerBarSeries, registerBoxplotSeries, registerCirclePackingSeries, registerCircularProgressSeries, registerDotSeries, registerFunnel3dSeries, registerFunnelSeries, registerGaugePointerSeries, registerGaugeSeries, registerHeatmapSeries, registerLineSeries, registerLinearProgressSeries, registerLinkSeries, registerMapSeries, registerPie3dSeries, registerPieSeries, registerRadarSeries, registerRangeAreaSeries, registerRangeColumn3dSeries, registerRangeColumnSeries, registerRoseSeries, registerSankeySeries, registerScatterSeries, registerSunBurstSeries, registerTreemapSeries, registerWaterfallSeries, registerWordCloud3dSeries, registerWordCloudSeries };

export * from "./interface";
