// @visactor/vchart-semi-theme/esm/dark/color-scheme.js
import { lightTheme } from "@visactor/vchart";

export const colorScheme = {
    default: {
        dataScheme: lightTheme.colorScheme.default.dataScheme,
        palette: {
            backgroundColor: "#16161a",
            borderColor: "rgba(255,255,255,0.08)",
            shadowColor: "rgba(0,0,0,0.25)",
            hoverBackgroundColor: "rgba(255,255,255,0.12)",
            sliderRailColor: "rgba(255,255,255,0.12)",
            sliderHandleColor: "#e4e7f5",
            sliderTrackColor: "rgba(84,169,255,1)",
            popupBackgroundColor: "#43444a",
            primaryFontColor: "rgba(249,249,249,1)",
            secondaryFontColor: "rgba(249,249,249,0.8)",
            tertiaryFontColor: "rgba(249,249,249,0.6)",
            axisLabelFontColor: "rgba(249,249,249,0.6)",
            disableFontColor: "rgba(249,249,249,0.35)",
            axisMarkerFontColor: "#16161a",
            axisGridColor: "rgba(255,255,255,0.08)",
            axisDomainColor: "rgba(255,255,255,0.08)",
            dataZoomHandleStrokeColor: "rgba(46,50,56,0.13)",
            dataZoomChartColor: "rgba(255,255,255,0.16)",
            playerControllerColor: "rgba(84,169,255,1)",
            axisMarkerBackgroundColor: "rgba(249,249,249,1)",
            markLabelBackgroundColor: "rgba(255,255,255,0.08)",
            markLineStrokeColor: "rgba(249,249,249,0.8)",
            dangerColor: "rgba(252,114,90,1)",
            warningColor: "rgba(255,174,67,1)",
            successColor: "rgba(93,194,100,1)",
            infoColor: "rgba(84,169,255,1)"
        }
    }
};
