// @visactor/vchart/esm/theme/builtin/common/series/index.js
import { area } from "./area";

import { bar } from "./bar";

import { bar3d } from "./bar3d";

import { circularProgress } from "./circular-progress";

import { dot } from "./dot";

import { funnel } from "./funnel";

import { funnel3d } from "./funnel3d";

import { gauge } from "./gauge";

import { gaugePointer } from "./gauge-pointer";

import { line } from "./line";

import { linearProgress } from "./linear-progress";

import { link } from "./link";

import { map } from "./map";

import { pie } from "./pie";

import { pie3d } from "./pie3d";

import { radar } from "./radar";

import { scatter } from "./scatter";

import { waterfall } from "./waterfall";

import { wordCloud } from "./word-cloud";

import { wordCloud3d } from "./word-cloud3d";

import { treemap } from "./treemap";

import { sunburst } from "./sunburst";

import { rangeColumn } from "./rangeColumn";

import { circlePacking } from "./circle-packing";

import { heatmap } from "./heatmap";

import { sankey } from "./sankey";

import { rose } from "./rose";

import { boxPlot } from "./box-plot";

import { correlation } from "./correlation";

export const series = {
    scatter: scatter,
    line: line,
    area: area,
    bar: bar,
    bar3d: bar3d,
    pie: pie,
    pie3d: pie3d,
    map: map,
    radar: radar,
    dot: dot,
    link: link,
    wordCloud: wordCloud,
    wordCloud3d: wordCloud3d,
    funnel: funnel,
    funnel3d: funnel3d,
    linearProgress: linearProgress,
    circularProgress: circularProgress,
    waterfall: waterfall,
    gauge: gauge,
    gaugePointer: gaugePointer,
    treemap: treemap,
    sunburst: sunburst,
    rangeColumn: rangeColumn,
    circlePacking: circlePacking,
    heatmap: heatmap,
    sankey: sankey,
    rose: rose,
    boxPlot: boxPlot,
    correlation: correlation
};
