// @visactor/vchart-semi-theme/cjs/generator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.generateVChartSemiTheme = void 0;

const dark_1 = require("./dark"), light_1 = require("./light"), token_map_1 = require("./common/token-map"), vchart_theme_utils_1 = require("@visactor/vchart-theme-utils"), baseThemeMap = {
    light: light_1.semiDesignLight,
    dark: dark_1.semiDesignDark
}, generateVChartSemiTheme = (mode, chartContainer) => {
    const baseTheme = baseThemeMap[mode], {dataScheme: dataScheme, palette: palette} = baseTheme.colorScheme.default, colorScheme = {
        default: {
            dataScheme: (0, vchart_theme_utils_1.generateDataScheme)(mode, token_map_1.dataSchemeTokenMap, dataScheme, chartContainer),
            palette: (0, vchart_theme_utils_1.generatePalette)(mode, token_map_1.paletteTokenMap, palette, chartContainer)
        }
    };
    return Object.assign(Object.assign({}, baseTheme), {
        colorScheme: colorScheme
    });
};

exports.generateVChartSemiTheme = generateVChartSemiTheme;
