// @visactor/vgrammar-core/cjs/graph/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.DefaultAnimationParameters = exports.DefaultAnimationControlOptions = exports.DefaultAnimationEasing = exports.DefaultAnimationOneByOne = exports.DefaultAnimationLoop = exports.DefaultAnimationDelayAfter = exports.DefaultAnimationDelay = exports.DefaultAnimationDuration = exports.DefaultAnimationStartTime = exports.ImmediateAnimationState = exports.DefaultEnableExitAnimation = exports.DefaultSplitPath = exports.DefaultMorphAll = exports.DefaultMorph = exports.DefaultReuse = exports.DefaultGroupKeys = exports.DefaultMarkData = exports.DefaultKey = exports.Mark3DType = exports.CollectionMarkType = exports.BridgeElementKey = void 0;

const enums_1 = require("./enums");

exports.BridgeElementKey = "__vgrammar_scene_item__", exports.CollectionMarkType = [ enums_1.GrammarMarkType.line, enums_1.GrammarMarkType.area ], 
exports.Mark3DType = [ enums_1.GrammarMarkType.arc3d, enums_1.GrammarMarkType.rect3d, enums_1.GrammarMarkType.pyramid3d ], 
exports.DefaultKey = "key", exports.DefaultMarkData = [ {} ], exports.DefaultGroupKeys = [ exports.DefaultKey ], 
exports.DefaultReuse = !0, exports.DefaultMorph = !0, exports.DefaultMorphAll = !1, 
exports.DefaultSplitPath = null, exports.DefaultEnableExitAnimation = !0, exports.ImmediateAnimationState = "VGRAMMAR_IMMEDIATE_ANIMATION", 
exports.DefaultAnimationStartTime = 0, exports.DefaultAnimationDuration = 1e3, exports.DefaultAnimationDelay = 0, 
exports.DefaultAnimationDelayAfter = 0, exports.DefaultAnimationLoop = !1, exports.DefaultAnimationOneByOne = !1, 
exports.DefaultAnimationEasing = "quintInOut", exports.DefaultAnimationControlOptions = {
    stopWhenStateChange: !1,
    immediatelyApply: !0
}, exports.DefaultAnimationParameters = "VGRAMMAR_ANIMATION_PARAMETERS";
