// @visactor/vchart/cjs/component/crosshair/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.LayoutType = exports.getDefaultCrosshairTriggerEventByMode = void 0;

const spec_1 = require("../../typings/spec"), util_1 = require("../../util");

function getDefaultCrosshairTriggerEventByMode(mode) {
    return mode === spec_1.RenderModeEnum["desktop-browser"] || mode === spec_1.RenderModeEnum["desktop-miniApp"] ? {
        click: "pointertap",
        hover: "pointermove",
        hoverOut: "pointerleave",
        clickOut: "pointerleave"
    } : (0, util_1.isMobileLikeMode)(mode) || (0, util_1.isMiniAppLikeMode)(mode) ? {
        click: "tap",
        hover: [ "pointerdown", "pointermove" ],
        hoverOut: "pointerleave",
        clickOut: "pointerleave"
    } : null;
}

var LayoutType;

exports.getDefaultCrosshairTriggerEventByMode = getDefaultCrosshairTriggerEventByMode, 
function(LayoutType) {
    LayoutType[LayoutType.ALL = 3] = "ALL", LayoutType[LayoutType.HORIZONTAL = 2] = "HORIZONTAL", 
    LayoutType[LayoutType.VERTICAL = 1] = "VERTICAL", LayoutType[LayoutType.NONE = 0] = "NONE";
}(LayoutType = exports.LayoutType || (exports.LayoutType = {}));
