// @visactor/vchart/cjs/component/axis/polar/linear-axis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerPolarLinearAxis = exports.PolarLinearAxis = void 0;

const vscale_1 = require("@visactor/vscale"), type_1 = require("../../interface/type"), axis_1 = require("./axis"), linear_axis_mixin_1 = require("../mixin/linear-axis-mixin"), vutils_1 = require("@visactor/vutils"), factory_1 = require("../../../core/factory"), base_axis_1 = require("../base-axis");

class PolarLinearAxis extends axis_1.PolarAxis {
    constructor() {
        super(...arguments), this.type = type_1.ComponentTypeEnum.polarLinearAxis, this._zero = !0, 
        this._nice = !0, this._scale = new vscale_1.LinearScale;
    }
    setAttrFromSpec() {
        super.setAttrFromSpec(), this.setExtraAttrFromSpec();
    }
    initScales() {
        super.initScales(), this.setScaleNice();
    }
    computeDomain(data) {
        return this.computeLinearDomain(data);
    }
}

exports.PolarLinearAxis = PolarLinearAxis, PolarLinearAxis.type = type_1.ComponentTypeEnum.polarLinearAxis, 
PolarLinearAxis.specKey = "axes", (0, vutils_1.mixin)(PolarLinearAxis, linear_axis_mixin_1.LinearAxisMixin);

const registerPolarLinearAxis = () => {
    (0, base_axis_1.registerAxis)(), factory_1.Factory.registerComponent(PolarLinearAxis.type, PolarLinearAxis);
};

exports.registerPolarLinearAxis = registerPolarLinearAxis;
