// @visactor/vchart/cjs/series/range-area/tooltip-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.RangeAreaSeriesTooltipHelper = void 0;

const tooltip_helper_1 = require("../base/tooltip-helper"), space_1 = require("../../typings/space");

class RangeAreaSeriesTooltipHelper extends tooltip_helper_1.BaseSeriesTooltipHelper {
    getDefaultTooltipPattern(activeType, dimensionInfo) {
        if ("mark" === activeType) return {
            visible: !0,
            activeType: activeType,
            title: {
                key: void 0,
                value: this.titleValueCallback,
                hasShape: !1
            },
            content: [ {
                key: this.contentKeyCallback,
                value: this.contentValueCallback,
                hasShape: !0,
                shapeType: this.contentShapeTypeCallback,
                shapeColor: this.contentShapeColorCallback,
                shapeStroke: this.contentShapeColorCallback,
                shapeHollow: !1
            } ]
        };
        if ("dimension" === activeType && dimensionInfo) {
            const title = {
                key: void 0,
                value: this._getDimensionData,
                hasShape: !1
            }, content = [];
            return dimensionInfo.forEach((({data: data}) => data.forEach((({series: series}) => {
                "rangeArea" === series.type && content.push({
                    seriesId: series.id,
                    key: this.contentKeyCallback,
                    value: datum => this.series.getSpec().direction === space_1.Direction.horizontal ? datum[this.series.getSpec().xField[0]] + "-" + datum[this.series.getSpec().xField[1]] : datum[this.series.getSpec().yField[0]] + "-" + datum[this.series.getSpec().yField[1]],
                    hasShape: !0,
                    shapeType: this.contentShapeTypeCallback,
                    shapeColor: this.contentShapeColorCallback,
                    shapeStroke: this.contentShapeColorCallback,
                    shapeHollow: !1
                });
            })))), {
                visible: !0,
                activeType: activeType,
                title: title,
                content: content
            };
        }
        return null;
    }
}

exports.RangeAreaSeriesTooltipHelper = RangeAreaSeriesTooltipHelper;
