// @visactor/vutils/cjs/dom.js
"use strict";

function getContainerSize(el, defaultWidth = 0, defaultHeight = 0) {
    if (!el) return {
        width: defaultWidth,
        height: defaultHeight
    };
    let getComputedStyle;
    try {
        getComputedStyle = null === window || void 0 === window ? void 0 : window.getComputedStyle;
    } catch (e) {
        getComputedStyle = () => ({});
    }
    const style = getComputedStyle(el);
    if (/^(\d*\.?\d+)(px)$/.exec(style.width)) {
        const computedWidth = parseFloat(style.width) - parseFloat(style.paddingLeft) - parseFloat(style.paddingRight) || el.clientWidth - 1, computedHeight = parseFloat(style.height) - parseFloat(style.paddingTop) - parseFloat(style.paddingBottom) || el.clientHeight - 1;
        return {
            width: computedWidth <= 0 ? defaultWidth : computedWidth,
            height: computedHeight <= 0 ? defaultHeight : computedHeight
        };
    }
    return {
        width: defaultWidth,
        height: defaultHeight
    };
}

function getElementAbsolutePosition(element) {
    const {x: x, y: y} = element.getBoundingClientRect();
    return {
        x: x,
        y: y
    };
}

function getElementRelativePosition(element, base) {
    const posElement = getElementAbsolutePosition(element), posBase = getElementAbsolutePosition(base);
    return {
        x: posElement.x - posBase.x,
        y: posElement.y - posBase.y
    };
}

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.hasParentElement = exports.getScale = exports.getScaleY = exports.getScaleX = exports.getScrollTop = exports.getScrollLeft = exports.getElementRelativePosition = exports.getElementAbsolutePosition = exports.getContainerSize = void 0, 
exports.getContainerSize = getContainerSize, exports.getElementAbsolutePosition = getElementAbsolutePosition, 
exports.getElementRelativePosition = getElementRelativePosition;

const getScrollLeft = element => {
    var _a, _b, _c;
    return element === (null === (_a = null === globalThis || void 0 === globalThis ? void 0 : globalThis.document) || void 0 === _a ? void 0 : _a.body) ? (null === (_c = null === (_b = null === globalThis || void 0 === globalThis ? void 0 : globalThis.document) || void 0 === _b ? void 0 : _b.documentElement) || void 0 === _c ? void 0 : _c.scrollLeft) || element.scrollLeft : "html" === element.tagName.toLowerCase() ? 0 : element.scrollLeft;
};

exports.getScrollLeft = getScrollLeft;

const getScrollTop = element => {
    var _a, _b, _c;
    return element === (null === (_a = null === globalThis || void 0 === globalThis ? void 0 : globalThis.document) || void 0 === _a ? void 0 : _a.body) ? (null === (_c = null === (_b = null === globalThis || void 0 === globalThis ? void 0 : globalThis.document) || void 0 === _b ? void 0 : _b.documentElement) || void 0 === _c ? void 0 : _c.scrollTop) || element.scrollTop : "html" === element.tagName.toLowerCase() ? 0 : element.scrollTop;
};

exports.getScrollTop = getScrollTop;

const getScaleX = element => element.getBoundingClientRect().width / element.offsetWidth;

exports.getScaleX = getScaleX;

const getScaleY = element => element.getBoundingClientRect().height / element.offsetHeight;

exports.getScaleY = getScaleY;

const getScale = element => element.offsetWidth > 0 ? (0, exports.getScaleX)(element) : (0, 
exports.getScaleY)(element);

function hasParentElement(element, target) {
    let parent = element.parentNode;
    for (;null !== parent; ) {
        if (parent === target) return !0;
        parent = parent.parentNode;
    }
    return !1;
}

exports.getScale = getScale, exports.hasParentElement = hasParentElement;
