// @visactor/vchart/cjs/component/tooltip/handler/dom/model/style-constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.defaultShapeStyle = exports.defaultValueStyle = exports.defaultAdaptiveKeyStyle = exports.defaultKeyStyle = exports.defaultContentColumnStyle = exports.defaultContentContainerStyle = exports.defaultContainerStyle = exports.defaultH2Style = void 0, 
exports.defaultH2Style = {
    fontSize: "13px",
    marginBottom: "0px",
    fontWeight: "normal"
}, exports.defaultContainerStyle = {
    boxSizing: "border-box"
}, exports.defaultContentContainerStyle = {
    display: "flex",
    whiteSpace: "nowrap"
}, exports.defaultContentColumnStyle = {
    display: "flex",
    flexDirection: "column",
    justifyContent: "space-around",
    fontSize: "0"
}, exports.defaultKeyStyle = {
    paddingTop: "0px",
    paddingBottom: "0px",
    textAlign: "left",
    fontWeight: "normal"
}, exports.defaultAdaptiveKeyStyle = {
    paddingTop: "0px",
    paddingBottom: "0px",
    textAlign: "left",
    fontWeight: "normal"
}, exports.defaultValueStyle = {
    paddingTop: "0px",
    paddingBottom: "0px",
    textAlign: "right",
    fontWeight: "normal"
}, exports.defaultShapeStyle = {
    lineHeight: "normal"
};
