// @visactor/vchart/cjs/component/tooltip/handler/dom/model/tooltip-model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.TooltipModel = void 0;

const vutils_1 = require("@visactor/vutils"), base_tooltip_model_1 = require("./base-tooltip-model"), content_model_1 = require("./content-model"), title_model_1 = require("./title-model"), style_constants_1 = require("./style-constants"), merge_spec_1 = require("../../../../../util/spec/merge-spec");

class TooltipModel extends base_tooltip_model_1.BaseTooltipModel {
    constructor(parent, option, classList, id) {
        super(parent, option, 0), this.title = null, this.content = null, this._classList = classList, 
        this._id = id;
    }
    setVisibility(visibility) {
        if (super.setVisibility(visibility), !this.product) return;
        const {classList: classList} = this.product;
        visibility ? classList.add("visible") : classList.remove("visible");
    }
    init() {
        var _a;
        const tooltipActual = this._option.getTooltipActual();
        this.product || this._initPanel(this._classList, this._id);
        const {title: title} = tooltipActual;
        !1 !== (null == title ? void 0 : title.visible) && (0, vutils_1.isValid)(null == title ? void 0 : title.value) ? this.title || this._initTitle() : this.title && this._releaseTitle();
        (null !== (_a = tooltipActual.content) && void 0 !== _a ? _a : []).length > 0 ? this.content || this._initContent() : this.content && this._releaseContent();
    }
    _initPanel(classList, id) {
        const panel = this.createElement("div", classList, {
            left: "0",
            top: "0",
            pointerEvents: "none",
            padding: "12px",
            position: "absolute",
            zIndex: "99999999999999",
            fontFamily: "sans-serif",
            fontSize: "11px",
            borderRadius: "3px",
            borderStyle: "solid",
            lineHeight: "initial",
            background: "#fff",
            boxShadow: "2px 2px 4px rgba(0, 0, 0, 0.1)",
            maxWidth: "100wh",
            maxHeight: "100vh"
        }, id);
        this.product = panel;
    }
    _initTitle() {
        const title = new title_model_1.TitleModel(this.product, this._option, 0);
        title.init(), this.title = title, this.children[title.childIndex] = title;
    }
    _releaseTitle() {
        this.title && (this.title.release(), delete this.children[this.title.childIndex], 
        this.title = null);
    }
    _initContent() {
        const content = new content_model_1.ContentModel(this.product, this._option, 1);
        content.init(), this.content = content, this.children[content.childIndex] = content;
    }
    _releaseContent() {
        this.content && (this.content.release(), delete this.children[this.content.childIndex], 
        this.content = null);
    }
    setStyle() {
        const tooltipStyle = this._option.getTooltipStyle();
        super.setStyle((0, merge_spec_1.mergeSpec)({}, style_constants_1.defaultContainerStyle, tooltipStyle.panel)), 
        Object.values(this.children).forEach((c => c.setStyle()));
    }
    setContent() {
        Object.values(this.children).forEach((c => {
            c.setContent();
        }));
    }
    release() {
        super.release(), this.title = null, this.content = null;
    }
}

exports.TooltipModel = TooltipModel;
