// @visactor/vrender-components/cjs/util/matrix.js
"use strict";

function scale(vector, scale) {
    return [ vector[0] * scale, vector[1] * scale ];
}

function length(vector) {
    const [x, y] = vector;
    return Math.sqrt(x * x + y * y);
}

function normalize(vector) {
    const [x, y] = vector;
    let len = x * x + y * y;
    return len > 0 && (len = 1 / Math.sqrt(len)), [ vector[0] * len, vector[1] * len ];
}

function angle(vector1, vector2) {
    const [x1, y1] = vector1, [x2, y2] = vector2, mag = Math.sqrt((x1 * x1 + y1 * y1) * (x2 * x2 + y2 * y2)), cosine = mag && (x1 * x2 + y1 * y2) / mag;
    return Math.acos(Math.min(Math.max(cosine, -1), 1));
}

function direction(v1, v2) {
    return v1[0] * v2[1] - v2[0] * v1[1];
}

function angleTo(v1, v2, direct) {
    const ang = angle(v1, v2), angleLargeThanPI = direction(v1, v2) >= 0;
    return direct ? angleLargeThanPI ? 2 * Math.PI - ang : ang : angleLargeThanPI ? ang : 2 * Math.PI - ang;
}

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.angleTo = exports.direction = exports.angle = exports.normalize = exports.length = exports.scale = void 0, 
exports.scale = scale, exports.length = length, exports.normalize = normalize, exports.angle = angle, 
exports.direction = direction, exports.angleTo = angleTo;
