// @visactor/vrender-core/cjs/render/contributions/render/contributions/circle-contribution-render.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.defaultCircleBackgroundRenderContribution = exports.defaultCircleTextureRenderContribution = exports.defaultCircleRenderContribution = exports.DefaultCircleRenderContribution = void 0;

const canvas_utils_1 = require("../../../../common/canvas-utils"), base_contribution_render_1 = require("./base-contribution-render"), enums_1 = require("../../../../common/enums"), base_texture_contribution_render_1 = require("./base-texture-contribution-render");

class DefaultCircleRenderContribution {
    constructor() {
        this.time = enums_1.BaseRenderContributionTime.afterFillStroke, this.useStyle = !0, 
        this.order = 0;
    }
    drawShape(circle, context, x, y, doFill, doStroke, fVisible, sVisible, circleAttribute, drawContext, fillCb, strokeCb) {
        const {outerBorder: outerBorder, innerBorder: innerBorder} = circle.attribute, doOuterBorder = outerBorder && !1 !== outerBorder.visible, doInnerBorder = innerBorder && !1 !== innerBorder.visible;
        if (!doOuterBorder && !doInnerBorder) return;
        const {radius: radius = circleAttribute.radius, startAngle: startAngle = circleAttribute.startAngle, endAngle: endAngle = circleAttribute.endAngle, opacity: opacity = circleAttribute.opacity, x: originX = circleAttribute.x, y: originY = circleAttribute.y, scaleX: scaleX = circleAttribute.scaleX, scaleY: scaleY = circleAttribute.scaleY} = circle.attribute, doStrokeOuter = !(!outerBorder || !outerBorder.stroke), doStrokeInner = !(!innerBorder || !innerBorder.stroke);
        if (doOuterBorder) {
            const {distance: distance = circleAttribute.outerBorder.distance} = outerBorder, dw = (0, 
            canvas_utils_1.getScaledStroke)(context, distance, context.dpr);
            if (context.beginPath(), context.arc(x, y, radius + dw, startAngle, endAngle), context.closePath(), 
            context.setShadowBlendStyle && context.setShadowBlendStyle(circle, circle.attribute, circleAttribute), 
            strokeCb) strokeCb(context, outerBorder, circleAttribute.outerBorder); else if (doStrokeOuter) {
                const lastOpacity = circleAttribute.outerBorder.opacity;
                circleAttribute.outerBorder.opacity = opacity, context.setStrokeStyle(circle, outerBorder, (originX - x) / scaleX, (originY - y) / scaleY, circleAttribute.outerBorder), 
                circleAttribute.outerBorder.opacity = lastOpacity, context.stroke();
            }
        }
        if (doInnerBorder) {
            const {distance: distance = circleAttribute.innerBorder.distance} = innerBorder, dw = (0, 
            canvas_utils_1.getScaledStroke)(context, distance, context.dpr);
            if (context.beginPath(), context.arc(x, y, radius - dw, startAngle, endAngle), context.closePath(), 
            context.setShadowBlendStyle && context.setShadowBlendStyle(circle, circle.attribute, circleAttribute), 
            strokeCb) strokeCb(context, innerBorder, circleAttribute.innerBorder); else if (doStrokeInner) {
                const lastOpacity = circleAttribute.innerBorder.opacity;
                circleAttribute.innerBorder.opacity = opacity, context.setStrokeStyle(circle, innerBorder, (originX - x) / scaleX, (originY - y) / scaleY, circleAttribute.innerBorder), 
                circleAttribute.innerBorder.opacity = lastOpacity, context.stroke();
            }
        }
    }
}

exports.DefaultCircleRenderContribution = DefaultCircleRenderContribution, exports.defaultCircleRenderContribution = new DefaultCircleRenderContribution, 
exports.defaultCircleTextureRenderContribution = base_texture_contribution_render_1.defaultBaseTextureRenderContribution, 
exports.defaultCircleBackgroundRenderContribution = base_contribution_render_1.defaultBaseBackgroundRenderContribution;
