// @visactor/vchart/esm/chart/word-cloud/word-cloud.js
import { SeriesTypeEnum } from "../../series/interface/type";

import { ChartTypeEnum } from "../interface/type";

import { BaseWordCloudChart } from "./base/base";

import { registerWordCloudSeries, registerWordCloudShapeSeries } from "../../series/word-cloud/word-cloud";

import { Factory } from "../../core/factory";

import { WordCloudChartSpecTransformer } from "./word-cloud-transformer";

export class WordCloudChart extends BaseWordCloudChart {
    constructor() {
        super(...arguments), this.transformerConstructor = WordCloudChartSpecTransformer, 
        this.type = ChartTypeEnum.wordCloud, this.seriesType = SeriesTypeEnum.wordCloud;
    }
}

WordCloudChart.type = ChartTypeEnum.wordCloud, WordCloudChart.seriesType = SeriesTypeEnum.wordCloud, 
WordCloudChart.view = "singleDefault", WordCloudChart.transformerConstructor = WordCloudChartSpecTransformer;

export const registerWordCloudChart = () => {
    registerWordCloudSeries(), Factory.registerChart(WordCloudChart.type, WordCloudChart);
};

export const registerWordCloudShapeChart = () => {
    registerWordCloudShapeSeries(), registerWordCloudSeries(), Factory.registerChart(WordCloudChart.type, WordCloudChart);
};
