// @visactor/vchart/esm/chart/word-cloud/word-cloud-transformer.js
import { BaseWordCloudChartSpecTransformer } from "./base";

export class WordCloudChartSpecTransformer extends BaseWordCloudChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        const series = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            nameField: spec.nameField,
            valueField: spec.valueField,
            seriesField: spec.seriesField,
            fontFamilyField: spec.fontFamilyField,
            fontWeightField: spec.fontWeightField,
            fontStyleField: spec.fontStyleField,
            colorHexField: spec.colorHexField,
            colorMode: spec.colorMode,
            colorList: spec.colorList,
            rotateAngles: spec.rotateAngles,
            fontWeightRange: spec.fontWeightRange,
            fontSizeRange: spec.fontSizeRange,
            maskShape: spec.maskShape,
            keepAspect: spec.keepAspect,
            random: spec.random,
            wordCloudConfig: spec.wordCloudConfig,
            wordCloudShapeConfig: spec.wordCloudShapeConfig,
            word: spec.word,
            fillingWord: spec.fillingWord
        }), seriesType = this.seriesType;
        return seriesType && (series.type = seriesType, series[seriesType] = spec[seriesType]), 
        series;
    }
}
