// @visactor/vrender-components/cjs/label/area.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.AreaLabel = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("./base"), util_1 = require("./util");

class AreaLabel extends base_1.LabelBase {
    constructor(attributes) {
        super((0, vutils_1.merge)({}, AreaLabel.defaultAttributes, attributes)), this.name = "line-label";
    }
    getGraphicBounds(graphic, point = {}, position = "end") {
        var _a;
        if ("area" !== graphic.type) return super.getGraphicBounds(graphic, point);
        const points = (null === (_a = null == graphic ? void 0 : graphic.attribute) || void 0 === _a ? void 0 : _a.points) || [ point ], index = "start" === position ? 0 : points.length - 1;
        return {
            x1: points[index].x,
            x2: points[index].x,
            y1: points[index].y,
            y2: points[index].y
        };
    }
    labeling(textBounds, graphicBounds, position = "end", offset = 0) {
        return (0, util_1.labelingLineOrArea)(textBounds, graphicBounds, position, offset);
    }
}

exports.AreaLabel = AreaLabel, AreaLabel.defaultAttributes = {
    textStyle: {
        fontSize: 12,
        fill: "#000",
        textAlign: "center",
        textBaseline: "middle",
        boundsPadding: [ -1, 0, -1, 0 ]
    },
    position: "end",
    offset: 6,
    pickable: !1
};
