// @visactor/vrender-components/cjs/label/polygon.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.PolygonLabel = void 0;

const vutils_1 = require("@visactor/vutils"), base_1 = require("./base");

class PolygonLabel extends base_1.LabelBase {
    constructor(attributes, options) {
        super((null == options ? void 0 : options.skipDefault) ? attributes : (0, vutils_1.merge)({}, PolygonLabel.defaultAttributes, attributes)), 
        this.name = "polygon-label";
    }
    labeling(textBounds, graphicBounds, position = "center", offset = 0) {
        if (!textBounds || !graphicBounds) return;
        const {x1: x1, x2: x2, y2: y2, y1: y1} = textBounds;
        return {
            x: (x1 + x2) / 2,
            y: (y1 + y2) / 2
        };
    }
}

exports.PolygonLabel = PolygonLabel, PolygonLabel.defaultAttributes = {
    textStyle: {
        fontSize: 12,
        fill: "#000",
        textAlign: "center",
        textBaseline: "middle",
        boundsPadding: [ -1, 0, -1, 0 ]
    },
    position: "center",
    offset: 6,
    pickable: !1
};
