// @visactor/vchart/cjs/chart/polar/rose-like/rose-like-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.RoseLikeChartSpecTransformer = void 0;

const polar_transformer_1 = require("../polar-transformer");

class RoseLikeChartSpecTransformer extends polar_transformer_1.PolarChartSpecTransformer {
    needAxes() {
        return !0;
    }
    _getDefaultSeriesSpec(spec) {
        return Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            categoryField: spec.categoryField || spec.angleField,
            valueField: spec.valueField || spec.radiusField
        });
    }
    transformSpec(spec) {
        var _a;
        if (super.transformSpec(spec), this.needAxes()) {
            spec.axes || (spec.axes = []);
            const haxAxes = {
                radius: !1,
                angle: !1
            };
            (null !== (_a = spec.axes) && void 0 !== _a ? _a : []).forEach((axis => {
                const {orient: orient} = axis;
                "radius" === orient && (haxAxes.radius = !0), "angle" === orient && (haxAxes.angle = !0);
            })), haxAxes.angle || spec.axes.push({
                orient: "angle"
            }), haxAxes.radius || spec.axes.push({
                orient: "radius"
            });
        }
    }
}

exports.RoseLikeChartSpecTransformer = RoseLikeChartSpecTransformer;
