// @visactor/vchart/cjs/util/spec/merge-theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.transformSeriesThemeToMerge = exports.transformColorSchemeToMerge = exports.mergeTheme = void 0;

const vutils_1 = require("@visactor/vutils"), util_1 = require("../../theme/color-scheme/util"), theme_1 = require("../../series/interface/theme"), merge_spec_1 = require("./merge-spec");

function mergeTheme(target, ...sources) {
    return (0, merge_spec_1.mergeSpec)(transformThemeToMerge(target), ...sources.map(transformThemeToMerge));
}

function transformThemeToMerge(theme) {
    if (!theme) return theme;
    const colorScheme = transformColorSchemeToMerge(theme.colorScheme), {series: series} = theme, {mark: markByType, markByName: markByName} = theme;
    let newSeriesTheme;
    return (markByType || markByName) && (newSeriesTheme = Object.keys(theme_1.seriesMarkInfoMap).reduce(((newSeriesTheme, key) => {
        var _a;
        const value = null !== (_a = null == series ? void 0 : series[key]) && void 0 !== _a ? _a : {};
        return newSeriesTheme[key] = transformSeriesThemeToMerge(value, key, markByType, markByName), 
        newSeriesTheme;
    }), {})), Object.assign({}, theme, {
        colorScheme: colorScheme,
        series: Object.assign({}, theme.series, newSeriesTheme)
    });
}

function transformColorSchemeToMerge(colorScheme) {
    return colorScheme && (colorScheme = Object.keys(colorScheme).reduce(((scheme, key) => {
        const value = colorScheme[key];
        return scheme[key] = (0, util_1.transformColorSchemeToStandardStruct)(value), scheme;
    }), {})), colorScheme;
}

function transformSeriesThemeToMerge(seriesTheme, seriesType, markByType, markByName) {
    if (!theme_1.seriesMarkInfoMap[seriesType]) return seriesTheme;
    const newTheme = {};
    return Object.values(theme_1.seriesMarkInfoMap[seriesType]).forEach((({type: type, name: name}) => {
        newTheme[name] = (0, merge_spec_1.mergeSpec)({}, null == markByType ? void 0 : markByType[(0, 
        vutils_1.array)(type)[0]], null == markByName ? void 0 : markByName[name], null == seriesTheme ? void 0 : seriesTheme[name]);
    })), Object.assign(Object.assign({}, seriesTheme), newTheme);
}

exports.mergeTheme = mergeTheme, exports.transformColorSchemeToMerge = transformColorSchemeToMerge, 
exports.transformSeriesThemeToMerge = transformSeriesThemeToMerge;
