// @visactor/vgrammar-core/cjs/interactions/element-active.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.ElementActive = void 0;

const enums_1 = require("../graph/enums"), base_1 = require("./base");

class ElementActive extends base_1.BaseInteraction {
    constructor(view, options) {
        super(view, options), this.type = ElementActive.type, this.handleStart = e => {
            e.element && this._marks && this._marks.includes(e.element.mark) && e.element.addState(this.options.state);
        }, this.handleReset = e => {
            e.element && this._marks && this._marks.includes(e.element.mark) && e.element.removeState(this.options.state);
        }, this.options = Object.assign({}, ElementActive.defaultOptions, options), this._marks = view.getMarksBySelector(this.options.selector);
    }
    getEvents() {
        return [ {
            type: this.options.trigger,
            handler: this.handleStart
        }, {
            type: this.options.resetTrigger,
            handler: this.handleReset
        } ];
    }
}

exports.ElementActive = ElementActive, ElementActive.type = "element-active", ElementActive.defaultOptions = {
    state: enums_1.InteractionStateEnum.active,
    trigger: "pointerover",
    resetTrigger: "pointerout"
};
