// @visactor/vgrammar-core/cjs/interactions/filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.Filter = exports.FilterMixin = void 0;

const base_1 = require("./base"), vutils_1 = require("@visactor/vutils");

class FilterMixin {
    _filterData(data, source, filterRank, getFilterValue, filter, transform) {
        const dataGrammar = (0, vutils_1.isString)(data) ? this.view.getDataById(data) : data;
        return dataGrammar && (this.handleFilter = event => {
            const element = null == event ? void 0 : event.element;
            dataGrammar && (!source || element && element.mark === source) && (getFilterValue && (this._filterValue = getFilterValue(event)), 
            dataGrammar.commit(), this.view.runAsync());
        }, this._dataFilter = {
            source: source ? `${source.uid}` : null,
            rank: filterRank,
            filter: data => {
                if (!this._filterValue) return data;
                const filteredData = filter ? data.filter((datum => filter(datum, this._filterValue))) : data;
                return transform ? transform(filteredData, this._filterValue) : filteredData;
            }
        }, dataGrammar.addDataFilter(this._dataFilter)), this;
    }
}

exports.FilterMixin = FilterMixin;

class Filter extends base_1.BaseInteraction {
    constructor(view, options) {
        super(view, options), options.target && (this._data = (0, vutils_1.isString)(options.target.data) ? view.getDataById(options.target.data) : options.target.data);
    }
}

exports.Filter = Filter, Filter.defaultOptions = {};
