// @visactor/vgrammar-core/cjs/interactions/scrollbar-filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.ScrollbarFilter = void 0;

const enums_1 = require("../graph/enums"), vutils_1 = require("@visactor/vutils"), filter_1 = require("./filter");

class ScrollbarFilter extends filter_1.Filter {
    constructor(view, options) {
        super(view, options), this.type = ScrollbarFilter.type, this.options = Object.assign({}, ScrollbarFilter.defaultOptions, options), 
        this._marks = view.getMarksBySelector(this.options.source).filter((mark => mark.markType === enums_1.GrammarMarkType.component && "scrollbar" === mark.componentType)), 
        this._data = (0, vutils_1.isString)(this.options.target.data) ? view.getDataById(this.options.target.data) : this.options.target.data;
    }
    getEvents() {
        if (!this._marks || 0 === this._marks.length) return [];
        const scrollbar = this._marks[0];
        if (!this._data || !scrollbar) return [];
        const filter = this.options.target.filter, transform = this.options.target.transform, dataFilter = (0, 
        vutils_1.isString)(filter) ? (datum, filterValue) => {
            const scale = scrollbar.getScale().getScale(), range = scale.range(), datumRatio = (scale.scale(datum[filter]) - range[0]) / (range[range.length - 1] - range[0]);
            return filterValue.startRatio <= datumRatio && filterValue.endRatio >= datumRatio;
        } : filter;
        return this._filterData(this._data, scrollbar, enums_1.DataFilterRank.scrollbar, (event => {
            if ((0, vutils_1.isString)(filter)) {
                const range = event.detail.value, scaleGrammar = scrollbar.getScale();
                if (scaleGrammar) {
                    const scale = scaleGrammar.getScale(), scaleRange = scale.range();
                    return {
                        start: scale.invert(range[0] * (scaleRange[1] - scaleRange[0]) + scaleRange[0]),
                        end: scale.invert(range[1] * (scaleRange[1] - scaleRange[0]) + scaleRange[0]),
                        startRatio: range[0],
                        endRatio: range[1]
                    };
                }
                return {
                    startRatio: range[0],
                    endRatio: range[1]
                };
            }
            return {
                startRatio: event.detail.value[0],
                endRatio: event.detail.value[1]
            };
        }), dataFilter, transform), [ {
            type: "scroll",
            handler: this.handleFilter
        } ];
    }
}

exports.ScrollbarFilter = ScrollbarFilter, ScrollbarFilter.type = "scrollbar-filter", 
ScrollbarFilter.defaultOptions = {};
