// @visactor/vrender-components/es/label/overlap/scaler.js
import { clamp as clampRange } from "@visactor/vutils";

import { bitmap } from "./bitmap";

export function bitmapTool(width, height, padding = 0) {
    const ratio = Math.max(1, Math.sqrt(width * height / 1e6)), w = ~~((width + 2 * padding + ratio) / ratio), h = ~~((height + 2 * padding + ratio) / ratio), scale = _ => ~~((_ + padding) / ratio);
    return scale.invert = _ => _ * ratio - padding, scale.bitmap = () => bitmap(w, h), 
    scale.ratio = ratio, scale.padding = padding, scale.width = width, scale.height = height, 
    scale;
}

export function boundToRange($, bound, clamp = !1) {
    if (clamp) {
        const {x1: x1, x2: x2, y1: y1, y2: y2} = bound, _x1 = clampRange(x1, 0, $.width), _x2 = clampRange(x2, 0, $.width), _y1 = clampRange(y1, 0, $.height), _y2 = clampRange(y2, 0, $.height);
        return {
            x1: $(_x1),
            x2: $(_x2),
            y1: $(_y1),
            y2: $(_y2)
        };
    }
    return {
        x1: $(bound.x1),
        x2: $(bound.x2),
        y1: $(bound.y1),
        y2: $(bound.y2)
    };
}
